{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From ChiP to Core Regulatory Circuitry.\n",
    "Jérémie Kalfon\n",
    "@jkobject\n",
    "\n",
    "[www.jkobject.com](https://www.jkobject.com)\n",
    "\n",
    "in partnership with the Broad Institute\n",
    "\n",
    "thanks to young Lab @ MIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named Bio",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-fd81831fa8ee>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mpath_to_rose\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/Users/jeremie/Documents/logiciels/rose\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_to_rose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mROSE_utils\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: No module named Bio"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "path_to_rose = \"/Users/jeremie/Documents/logiciels/rose\"\n",
    "os.chdir(path_to_rose)\n",
    "import ROSE_utils\n",
    "os.chdir(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### you might need to install an extention to curl to download from an sftp server. otherwise reinstall the latest version of curl with libssh\n",
    "\n",
    "on mac: \n",
    "`brew install curl --with-libssh2=/usr/local`\n",
    "after installing [conda]( \n",
    "```\n",
    "conda config --add channels defaults\n",
    "conda config --add channels bioconda\n",
    "conda config --add channels conda-forge\n",
    "```\n",
    "then\n",
    "`conda config install bowtie2 && pip install MACS2`\n",
    "then `git clone` and provide path to [ROSE](https://bitbucket.org/young_computation/rose/src) and [CRCMapper](https://bitbucket.org/young_computation/crcmapper)\n",
    "\n",
    "for CRC mapper you need [meme](http://meme-suite.org/doc/install.html?man_type=web)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7567143977"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mypath= \"datatosend\"\n",
    "! curl -k \"sftp://bioinfodata.wistar.upenn.edu/datatosend.tar.gz\" --user [NAME] > $mypath.tar.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## the call might stop for many reasons. in this case just run this function below to continue it from where you started as the -R option is not working well with curl and sftp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def notfinished():\n",
    "    s = os.path.getsize(mypath+'.tar.gz')\n",
    "    ! curl -C $s -k \"sftp://bioinfodata.wistar.upenn.edu/datatosend.tar.gz\" --user [NAME] > $mypath1.tar.gz\n",
    "    ! cat $mypath.tar.gz $mypath1.tar.gz > $mypath2.tar.gz\n",
    "    ! rm $mypath.tar.gz $mypath1.tar.gz\n",
    "    ! mv $mypath2.tar.gz $mypath.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### here we prepare the libs and the alignments with the hg19/37 version of the genome as it is the best one currently working with CRCmapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! gzip -d $mypath.tar.gz\n",
    "! curl ftp://ftp.ensembl.org/pub/grch37/current/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.dna.toplevel.fa.gz > reference.fa.gz\n",
    "! gzip -d reference.fa.gz\n",
    "! mkdir reference\n",
    "os.chdir(\"reference\")\n",
    "! wget --timestamping 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chr{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y}.fa.gz'\n",
    "! gzip -d *.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### some visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1233:1928\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1233:1928\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1233:1928 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('NTTCCAATGATCTCACATTAAACTGAAGCCTCTTTCCTCTGACAAACGTTTCCG...GAN', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1408:1913\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1408:1913\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1408:1913 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('NCTAATGTTTTTCTATTTTTAATAGAGATGTGGTTTCACCATGTTGGTCAGGCT...CCC', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1276:1938\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1276:1938\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1276:1938 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('NAGTCACGTGACCCAGAAACTTGTCCAAGGAGGCATGCATCTCTGGGCTGAAGT...GAN', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1494:1973\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1494:1973\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1494:1973 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('AGTTAAGTGCCACTGTTTTACAATTAAAAAGGGGATCATTTTAAATTTAAAAGC...TCT', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1375:1975\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1375:1975\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1375:1975 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('CAGGTTCACACTGCCACACCCAGCTAATTTTTTGTGTGTGTATTTTAGTAGAGA...CTA', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1537:1946\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1537:1946\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1537:1946 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('CAAAATACAAGAGTCATCCCTCGTGTCTCCCACCAAGGAAGTCATCAGACTGCA...GGT', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1672:1951\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1672:1951\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1672:1951 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('AGCAATTGCAACAAAAACAAAACTTGACATGTAGGACCTAATTTAACTAAAGAG...CTA', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1610:1964\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1610:1964\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1610:1964 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('GGAGTCTGGTCTTTTCCTCAGCACCATGAACAGTAAGGCTGCTTTGATATGGGT...TTG', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1536:1977\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1536:1977\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1536:1977 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('ACACCACTTTCTAGAGTTTGTCTTCTTAGACTGTGGCTTCTCATACTTTAATGA...CTG', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1650:1999\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1650:1999\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1650:1999 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('CAACTCACCTCTCCTGGCCTACAGGAACAAAGGGAAAGAGGTTGACATTATTTA...GAC', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1991:1991\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1991:1991\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1991:1991 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('GCACAGTGGCTCACGTCTGTAATCCCAGCACTTTGGGAGGCTGAGGTGGGAGGA...GAA', SingleLetterAlphabet())\n",
      "ID: 7001405:985:HJ7YWBCX2:1:1102:1807:1996\n",
      "Name: 7001405:985:HJ7YWBCX2:1:1102:1807:1996\n",
      "Description: 7001405:985:HJ7YWBCX2:1:1102:1807:1996 1:N:0:GCCAATGT\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('TCTTCTTCTCCCCCCTCCTCCCCTCCTTTTCCTTCTCCTCCTCCTCTTCAGGTA...TTA', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "for i, rec in enumerate(SeqIO.parse(mypath+\"/54-mp457-PDX34-H3K27ac-r2_S7_R1.fastq\", \"fastq\")):\n",
    "    print(rec)\n",
    "    if i > 10:\n",
    "        break    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## here we can see the -logp that the base call was incorrect (called phred quality/sanger quality...) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGTANCTCTGCTTAGGAATGAGAATGGTATTTGACTTGGTTTTAANGNATTAAANAGGTTTTTCTACCANCANA\n",
      "AAATGNACTGCTTGTTCTACACCACTTATGTTTTTTTTTATAGTGANCNTTATAANCTAAGAGAAACAGCNGANT\n",
      "GAAGTNGATTTTGCTTGTATATTTTGAGAGTAGGAGTAGCTTTGCTNCNTAAAACNCATACCTAAATCCTNCAGA\n",
      "GTTGTNTTAAGTGAAAGAGATAATGCATATAACAGTAGTGACTATANTNATTGAGCACTTACTATGTGACNTGCA\n",
      "TATGCNCAAGTATGTGTGTGTGCTAGCCAATGTGCAGAGAAGCTTTNCNTCCTAAATGGGACATCTTCATNAAAT\n",
      "ATGCTNTGAAGAAGCAAGCTGCAGTGTTGGAGAGGCCCATGTGGCANGGAAATGAGGGTGGCCTCCAGTCAATGG\n",
      "CTCACNGTGTTACCAGGTTGGAATGCAGTGGCGCGATCTCGGCTCACTGCAACCTCTGCCTCTTGGGTTCAAGTA\n",
      "AGTTTNTGGCTTTTACTACAATATATGTAGAACTGGATATTTAGAATATCATTAAAACAAAGTACTTTGTGATTC\n",
      "TAGGANTTGTTTGCATGTGGAGTCAGATTCTCTGAGAACAAATGCCCATAATTTTGTGTATTTGGGCATAAAACA\n",
      "CTACTNTGAAATTTTGTATAGATAGGGTTAAATTTAGGAATATTTTCAAGACTTTGAATTACATAGAGTATTATC\n",
      "GTGCCNATTGGTGTATTTACAATCCCCGAGCTAGACACAAAGGTTCTCCAAGGCCCCACCAGACTCAGGAGCCCA\n",
      "TGTGANGTGTTCGTTCATCTCACAGAGTTTAACCTTTCTTTTCATAGAGCAGTTAGGAAACAGTCTGTTTGTAAA\n",
      "CAAAANAATGGAATCCAAAATCATTGACTGGAAAGGCTGGGTGTCGAAAGGAATTGACTCCAATGGAATGGAATC\n",
      "TGTCTNTCACAAAGGTACTTATTTTTTAATAATAATATCTTAATTAAAGATAGAAAGTCACAAGACAAACATGGC\n",
      "TGTTTNCTGTTTGCAGTTAAAAATGAGAGCTCTCAGTCCTGCTTTGTTGGGGCCTGTGTGTGTATGTGTGTCCTG\n",
      "GTAGTNGCTCTCCTGCTGCCCCAGGCTTGAAAGTGGGGAAAGCACAATTCTAGCCCTGTTGTTGGGGTACTTTCC\n",
      "TTAGCNTGGCTGGAAGCTCATGCATGAGTTTTGGAGTAAGACTTGGGTTGAAACCTTAGCTCTGCCGACTTATTA\n",
      "TGAAANATGAGAAATGCACACTGTAGGACCTGGAATATGGCGAGAAATCTGAAAATCACGGAAAATGAGAAATAC\n",
      "ATGAANTCTTGACAACACCAAGAGAAGAAAGGTTATACTTAATAAGTGATGTGCTTGTCAAGTTAACAAAGGGTC\n",
      "CTGAANGTCTTTTTCTATTTTTGTATTTTGTATTAGTTTTTTGTATTATTTTGTATTGATTTTTAAGAGCTTGGT\n",
      "TGGAANGGAATGGAATGGAATAGACTCGAATGAAATGGAGTGGACTCGAATGGAATGGAGTGGAATGGAATTGAC\n",
      "CTTTTNATATCTTTTTCAATTACTGTCTTCAAAGACTCAAAGTTTTTAGCATTAAAATTTTCACTTGCTTGAAGT\n",
      "CACTANATGAATGAGAGTGGTTGAAGCAGAATTCACATGAATCATCTCTCGAATCAACATAGTATGATAAATTGC\n",
      "CACATNAAAATGAAATAATCTCAGGGCTGAAGGGTTAAAGTACTAGTTACTCTGCCAGAGGACCTGGATTCAGTT\n",
      "CCAGGCTGGTCTCGAACTCCTGATCTCGTGATCTGCCTTGCCTCTGCCTCCCAAAGTGCTGGGATTACAGGCGTG\n",
      "GCATTCAGAGGAGCTTGGGATCCTCTTTCTGGCCAGCCCAGCTCCTCCAAGGAAGGGAAGTTTAACGATGTGAGG\n",
      "CTTCAGGGAGGGCTGAAACCTACAGGATCCCAATGTTGACCTGTTAGACGCTGGCTCCAGCTCCCCCGCCACTGT\n",
      "GTATAAGACTTTCATAGCTTGGCCACCCTCATCAATTTGGAGGGAACAAAGGGCCAAAGTAGGAAAACTATTTTT\n",
      "AAGATTGAACTCATGGCCTTGCTCAAACTAGGTGTGAATAGCTGTGCTGCGTATCTGAAATGCAAATTAAGCTAG\n",
      "CGTTCCATTTCATTGCATTCCATTCTATTCAACTCTATTGCCTTCCATTCCATTACATTCGATTACATTCCATGC\n",
      "GAAAAAAAGAGAAGAGATAACATAAGACAAGATAGGAAAGATGCTAATCAAGTTGAAAAGATACCTCAGAACGAA\n",
      "TTGCAGATTCTACACAAAGAGAGTTTCCAAACTGCTCTGTCAAAGGGAATGTTCAACTCTGTGACTTGAATGCAA\n",
      "CATATTTTGGTGCATTTAATGATGTCCAACATTTCTCTAATAATCTGTTCATATTCATTATTTTCTTTTTTTTTC\n",
      "AGATGGAAAACAATGATCTTCATACTCGGCCATTCATAGATCTAATCTCATACCTCTTCCTCAGAGCTTCTTGTC\n",
      "TGATGATGATCACGGCTACTACTACTACCTCTACTATGACTGTGACTTGTGATTTCTGGTATCATAGTACATAGT\n",
      "TTGGATGTGGACCTATCTTTTGGGGAGGTACCATTCAGCCTACTGTACCTGTTAATGGAATGGTTTGTTACATGG\n",
      "AGATACCTAAATAGTTTTTGAAAACCTCTTCTTCCGTGAGACTGGCTTGTTCCATTTGAGGAGTATGGGCATAGG\n",
      "AATTTTTGTATCTTTAGTAGAGATGGGGTTTCACCATGTTGGTCAGGCTGGTCTTGAATTCCTGACCTCAAGTGA\n",
      "AATCTCATCCCCATCCTGAGCGGCCTCTTTTCTAGGATCAAGATGAACACACTGCAGCTGAGGACAAGAGCCCCA\n",
      "ATTTAGATCGCTGGCTTTCTTCTGAACACCACAGTTGTGTGGGACTTTGGGCAATAATGGTGATCGTTGTTGAGG\n",
      "AGTCTTTAAGTTAGCTATTACAAATACATTCAAAGAACTAAATGAAACCATGTCTAAAGAATTAAAGGGAAGCAT\n",
      "TTGATGTGTTTTTAGTTTGGCAATAAACTTAAAATGAGTATAAATGAACCTACTAAACACTGTCACTATTGGGCA\n",
      "ATGCAGGTTCTTATAGGCATCTTCCCTATCTGACAATGGGTGTAATATTTTGTCCCTCAGACCTTTAGAAAAAAA\n",
      "AGATAAGATTCCGGCTTCCATGGAAGGCTGCAGTGCAACAGGCAACTTCAGCCCCGTGTCATACAGGCCCTTGAG\n",
      "GTTATTATTTTATTATTACTATTCATTGTTTATCTGAAATTCAGAACTAACTGGGCATCTTGTATTTTTATTTGT\n",
      "ATCCTCAGACAATGATGTGTCCTGCCTGCTCAAGGACCACTTCTTGAACTGGAAGGAGCCCTCTCTGGGACTTGG\n",
      "AGCATCTGTTTTGAATATGATGCTTCCTTTCAAGGATATGACCTGAATCGTCAATCACATTCTCATTATAATCAT\n",
      "TTTCTTTCTTTTCTTTCTTCTTTTTTCCTTTTTTTCTTCTGTCTTTCTTCTTTCTTTTTTCCTTTTTTTCTTCTT\n",
      "AACATTGAAATACTAAGGGACAAGAAGGAATTTTTCAACAAACAGGAAAACACATAAAGCTCTTAGATACAAAAA\n",
      "AAGTCATGAAAATTGATGAAAATCTACAACTGCCTCTTTACCAAAGCTGCATAATTTCAAACCACATTATAAATA\n",
      "GACAAAGCACATGTTTCAGAGAGCACGGGGTTGGGGGTAAGGTTATAGATTAACAGCATCCCAAGGCAGAAGAAT\n",
      "GTGGTTTTATCTACCTTTGGTCTTTGATGATGGTGACGTACAGATGGGGTTTTGGTGTGGATGTCCTTTCTGTTT\n",
      "CCTAACTTATTCACTGTTATGTTAATTGCATCTATGGCAGTATCTATTAAGTCTTAGGCATTGGGAATAAAGATT\n",
      "CACAGCCTAGCTCTAGAGCCTATAGCTAGGATTCCTTTATGGTGGGGCCATGCGGTCCCTTTCAAAGGGTAATGA\n",
      "TCCTACTGTCACATTGTTCTAATTGACCAGTGTGGTTTTTAAATAAGAGTGCAGAATATAAAGTATGAGGTGGAA\n",
      "CCCCTTTTGATCCTTGTCATCATAGAGTTTACTGTGTGTTTAGACCGTTCTTGCATTGCTATAAAGAAATACCTG\n",
      "TTTTCCACCAACATGCATTGACATTACCCCGCAGGTATTTTACCAGAAAATAAGATCTAATAAATTATACGTATT\n",
      "GCCCCAGGAATGATACAAAATGATGCCAGGTCGCAGAAGCCCTTGGCCTCAATGGGGAGGAAAAATGGATAAGAA\n",
      "GCCACACGAGCTGTCCTTGATCTTGCCTTTACTCTCAATCTGTAGGACCTTAAGGCTGGCTCTCCATTAGGCTCC\n",
      "CATCAAATGGAATCGAATGGAATCAACATCAAATGGAATCAAATGGAATCATGGAACGGAAATGAATGGAATCGG\n",
      "ATTCATTGATTTTTTTGAAGGTTTTTTTGTGGATCTGTCTCTTTCAGTTCTGCTCTGATCTCAGTTATTTCTTGC\n",
      "AGTCTTGGCTGCTGAGGCCTAGGCCAGGGTCCCTGTGGGTGGAGTGGGGCCTCAGCCCCTGCAGGGAGCTCTTTG\n",
      "ATCTCCAGATCATAAAATGATCATCTATTTGGCTCTTAAATATTCATATAAAAATTAACCCTAACCCTTGTCTTA\n",
      "GTTAGGGTGTCAATTTTAGTTCTTTTCTGCTTTCTCTTGTGGGCACTTAGTGCTATAAATTTCCCTCTACACACT\n",
      "CTGTTGAAGAAGCACCTCTCAGCCCCACATAGACAAGGTGCCACAAATGGTTTCCTGTTTTCTGAATTTAGAATG\n",
      "TTAAGATGGAGTCTTGCTCTGTCACCTAGGCTGCAGTACAGTGGTGCAATCATGGCTCACTGCAGCCTCAACCTC\n",
      "AGGATGAGACTGTCTCCAAAAAAGAAAAAAAAAAAATGATACCTAGAGGGGACAGTCCAATCGAAAGCCTCTCTT\n",
      "CTGGGTGTCATGGCAGGTGCCTATAGTCTCAGCTACTCGGGAGGCTGAGGCACAAGAGTTGCTTAACCCGGGAGG\n",
      "CCGAAAGGCTGATGGCCATGCCAGGTCAGGATGCGGCAGGCAGCCCCAGGGCGCTGTCCAGGTGGGGTTCCATTC\n",
      "AGGCTTCAGATTTCTCCAAGGGCCATTTCTAACAATAAGCACAGTTTAGTGGCTTCCTCTTATTCAGAGACTATA\n",
      "TGACCTCAGGTGATCTGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACCGGCGTGAGCCACCACGCCCCACCAA\n",
      "GTCCCTGCAACAGTTTGCTGAGAATGATGGTTTCCAGCTTCATCCATGTCCCTACAAAGGACATGAACTCATCAT\n",
      "CTTGGGCATGGCAGTGAAAGCAGACTTAGGGTATAATGAGACACTAGCCATTTTGATTGGGAGGTATAGGGGGCC\n",
      "GCTAGGCATTTTCTTACTTTCCCATTACTCGTGGGAATATTTTAAACAGAAGACAAGCACTCTTACCATGGAGCT\n",
      "ATGACTATGTCTTGGGGATGGTCATCTTGTAGAGTATCTTAGAGTTTCTCTGCATTTCCTGAATTTGAATGTCGA\n",
      "TAAAAATTATACTTCTTAAAATTTTTTTTTACCTAGTTATAAAAGTCACATAAACTTATTGAAAATATTTTGGAC\n",
      "ACAAACACCTCGTTGCTCAGCATAACCCTCTGTGAGACTCACTTGTAGCTCTGCTTTAAAAAAGGGTAAAAATGA\n",
      "AAAACTGAAAATCATGGAAAATGAGAAACATCCACTTTACGACTTGAAAAATGACGAAATCACTAAAAAACGTGA\n",
      "CAACTCCCTACTCATTCTTCTCTGGCAATACTATCACAGAGGATGGAGGGAATGATGGCAATATGAGAGTGGAAA\n",
      "TGTATTATATAGCTACATGTTTGATGAGGAAGCTGAGAGTCAGAGTTAAGTAATTTGCTCAAGTTTACACTGTTA\n",
      "TAGACAAAAACCCCATCACCTGGATGATCTGTACAGGATTATGTCACAAAGTTAGTGTAGGCAGATCCTAGACAA\n",
      "GGAAATCTTCATTCTCGGCAAATTAACACAGAAACAGAAAACCAAACACCACATGTTCTCACTCATAAGTGGGAG\n",
      "CATCTATGTGTAAACAGTCAAATGCACATATATAACTTTAAATTATCGTTTCCCTAGTTATTTATGTGGAGAAAC\n",
      "ATCATGACCATGTGGGGTTTATCCTAGGAATGCAAGGCTGTTTCAACATTTAAATATCAATCATTGTAAGTCACC\n",
      "TTCCATTAGGTTAAATTTGATGATGTTTCCTTTGAATTCCATTTCATGACAATTCCATTCAATACCAATTGATGA\n",
      "ATTTCCCAACAAAGCTGTGGATGGAAGTATTGTAATTGACAATGGTTTGGTCCAGGTATGGTCTTGTGGGTGCTT\n",
      "CACTCCAGCCTGGGTGACATAGCGAGACTCTGTCTCAAAAATAAATAAACATATCTAAATATAGAAAAGGTACAG\n",
      "AGCAACAATTTCATTCCTATAACAAGGGAATGATATTTCTGGTTTTTCCACATCGTTGTTAATAATTTATAGTGT\n",
      "TTATTCCATTCGATGATGATTCCATTCCATTCCATTTGATGATTCCATTCCATTCCATTAGATGTTGATTCCATT\n",
      "ATGTTAACCGGCAGAAGTATTTTCATCTTTGTCATACGAAAAAGAGTAATTTGCTCATAATAAGCCTTCAATATT\n",
      "TTAAAGGACCAATAAACAAAACCTTTATGTGCCATATAAATGTTGGCTAAGTTCTTTTACATTTTTCCCTCTTTT\n",
      "GTTTTACAGAAATGAGTCATTGTGCCTGGTCTGAGATATAAACATGATTGAAAGGGATTATGAAACTGGCTGCTG\n",
      "CACAGTGGGTGCCTCTGTAGGTTCCCCACCTTCAGTCACCTCCTCTAGGAGGTGGCGAAGGGCAGCCTGTTTGGC\n",
      "CTGGCTAAAGGCAATTATGGAAAAAATCTATAACTACCTTTATACTTAATGGTGAAATACTAAATGCTTCCCCCT\n",
      "CACTATTTCGATACTTCTTCATTTAGCAATGTACAGAGAGTCTCAACAGAGAAATTTTCCTGGCCACATTTTCAC\n",
      "AAAGGGAGGCAAAAGGCAAGGCTCCTAGGGACAGAGAGTGGTGACATGCCAGCTTCTGTTCGTTAAAACCCTGCT\n",
      "CAATGATACCTTGGTCAAGGAAGGAATAAAGAAAGAAATTAAAGACTTTTTAGAGTTTAATGAAAATGAAGCCAC\n",
      "CCTCGACTGAAGCTCTGTACTCATTAAACATGAAGTCCCCATTCCCCTCCCCCCCGCCCCTGGGACCCCCCATTC\n",
      "CAGAAATCTGAGATACTTCAGAGAATTGTGACAGTGGTGAGTGTAAGCATACACCTCTGCCATCTGCTGATATGC\n",
      "AGCATGCAGAGAAATAGCCATAACATTTCAGGATCATTGGAAGGGAAAAGCATTTTTCCTTGGTTATATGATAGT\n",
      "AGGCCCCAGTTTTCAAAACAAAACCTTCTGGGACTCCTGCAGCCTCAAAGCCCAATCCTTTAGGCAACACCAGCC\n",
      "GATTAAGTACAGAGTTTATTTAAGCCACAAACTTGAGAATGGCCACCTGGAAGCATTGATTGGAGCTGCCCTGAA\n"
     ]
    }
   ],
   "source": [
    "for i, rec in enumerate(SeqIO.parse(mypath+\"/54-mp457-PDX34-H3K27ac-r2_S7_R1.fastq\", \"fastq\")):\n",
    "    print(rec.seq)\n",
    "    if i > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: MACS2 callpeak [-h] -t TFILE [TFILE ...] [-c [CFILE [CFILE ...]]]\r\n",
      "                      [-f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}]\r\n",
      "                      [-g GSIZE] [-s TSIZE] [--keep-dup KEEPDUPLICATES]\r\n",
      "                      [--outdir OUTDIR] [-n NAME] [-B] [--verbose VERBOSE]\r\n",
      "                      [--trackline] [--SPMR] [--nomodel] [--shift SHIFT]\r\n",
      "                      [--extsize EXTSIZE] [--bw BW] [-m MFOLD MFOLD]\r\n",
      "                      [--fix-bimodal] [-q QVALUE | -p PVALUE]\r\n",
      "                      [--scale-to {large,small}] [--ratio RATIO]\r\n",
      "                      [--down-sample] [--seed SEED] [--tempdir TEMPDIR]\r\n",
      "                      [--nolambda] [--slocal SMALLLOCAL] [--llocal LARGELOCAL]\r\n",
      "                      [--max-gap MAXGAP] [--min-length MINLEN] [--broad]\r\n",
      "                      [--broad-cutoff BROADCUTOFF] [--cutoff-analysis]\r\n",
      "                      [--call-summits] [--fe-cutoff FECUTOFF]\r\n",
      "                      [--buffer-size BUFFER_SIZE] [--to-large]\r\n",
      "\r\n",
      "optional arguments:\r\n",
      "  -h, --help            show this help message and exit\r\n",
      "\r\n",
      "Input files arguments:\r\n",
      "  -t TFILE [TFILE ...], --treatment TFILE [TFILE ...]\r\n",
      "                        ChIP-seq treatment file. If multiple files are given\r\n",
      "                        as '-t A B C', then they will all be read and pooled\r\n",
      "                        together. REQUIRED.\r\n",
      "  -c [CFILE [CFILE ...]], --control [CFILE [CFILE ...]]\r\n",
      "                        Control file. If multiple files are given as '-c A B\r\n",
      "                        C', they will be pooled to estimate ChIP-seq\r\n",
      "                        background noise.\r\n",
      "  -f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}, --format {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}\r\n",
      "                        Format of tag file, \"AUTO\", \"BED\" or \"ELAND\" or\r\n",
      "                        \"ELANDMULTI\" or \"ELANDEXPORT\" or \"SAM\" or \"BAM\" or\r\n",
      "                        \"BOWTIE\" or \"BAMPE\" or \"BEDPE\". The default AUTO\r\n",
      "                        option will let MACS decide which format (except for\r\n",
      "                        BAMPE and BEDPE which should be implicitly set) the\r\n",
      "                        file is. Please check the definition in README. Please\r\n",
      "                        note that if the format is set as BAMPE or BEDPE,\r\n",
      "                        MACS2 will call its special Paired-end mode to call\r\n",
      "                        peaks by piling up the actual ChIPed fragments defined\r\n",
      "                        by both aligned ends, instead of predicting the\r\n",
      "                        fragment size first and extending reads. Also please\r\n",
      "                        note that the BEDPE only contains three columns, and\r\n",
      "                        is NOT the same BEDPE format used by BEDTOOLS.\r\n",
      "                        DEFAULT: \"AUTO\"\r\n",
      "  -g GSIZE, --gsize GSIZE\r\n",
      "                        Effective genome size. It can be 1.0e+9 or 1000000000,\r\n",
      "                        or shortcuts:'hs' for human (2.7e9), 'mm' for mouse\r\n",
      "                        (1.87e9), 'ce' for C. elegans (9e7) and 'dm' for\r\n",
      "                        fruitfly (1.2e8), Default:hs\r\n",
      "  -s TSIZE, --tsize TSIZE\r\n",
      "                        Tag size/read length. This will override the auto\r\n",
      "                        detected tag size. DEFAULT: Not set\r\n",
      "  --keep-dup KEEPDUPLICATES\r\n",
      "                        It controls the behavior towards duplicate tags at the\r\n",
      "                        exact same location -- the same coordination and the\r\n",
      "                        same strand. The 'auto' option makes MACS calculate\r\n",
      "                        the maximum tags at the exact same location based on\r\n",
      "                        binomal distribution using 1e-5 as pvalue cutoff; and\r\n",
      "                        the 'all' option keeps every tags. If an integer is\r\n",
      "                        given, at most this number of tags will be kept at the\r\n",
      "                        same location. Note, if you've used samtools or picard\r\n",
      "                        to flag reads as 'PCR/Optical duplicate' in bit 1024,\r\n",
      "                        MACS2 will still read them although the reads may be\r\n",
      "                        decided by MACS2 as duplicate later. If you plan to\r\n",
      "                        rely on samtools/picard/any other tool to filter\r\n",
      "                        duplicates, please remove those duplicate reads and\r\n",
      "                        save a new alignment file then ask MACS2 to keep all\r\n",
      "                        by '--keep-dup all'. The default is to keep one tag at\r\n",
      "                        the same location. Default: 1\r\n",
      "\r\n",
      "Output arguments:\r\n",
      "  --outdir OUTDIR       If specified all output files will be written to that\r\n",
      "                        directory. Default: the current working directory\r\n",
      "  -n NAME, --name NAME  Experiment name, which will be used to generate output\r\n",
      "                        file names. DEFAULT: \"NA\"\r\n",
      "  -B, --bdg             Whether or not to save extended fragment pileup, and\r\n",
      "                        local lambda tracks (two files) at every bp into a\r\n",
      "                        bedGraph file. DEFAULT: False\r\n",
      "  --verbose VERBOSE     Set verbose level of runtime message. 0: only show\r\n",
      "                        critical message, 1: show additional warning message,\r\n",
      "                        2: show process information, 3: show debug messages.\r\n",
      "                        DEFAULT:2\r\n",
      "  --trackline           Tells MACS to include trackline with bedGraph files.\r\n",
      "                        To include this trackline while displaying bedGraph at\r\n",
      "                        UCSC genome browser, can show name and description of\r\n",
      "                        the file as well. However my suggestion is to convert\r\n",
      "                        bedGraph to bigWig, then show the smaller and faster\r\n",
      "                        binary bigWig file at UCSC genome browser, as well as\r\n",
      "                        downstream analysis. Require -B to be set. Default:\r\n",
      "                        Not include trackline.\r\n",
      "  --SPMR                If True, MACS will SAVE signal per million reads for\r\n",
      "                        fragment pileup profiles. It won't interfere with\r\n",
      "                        computing pvalue/qvalue during peak calling, since\r\n",
      "                        internally MACS2 keeps using the raw pileup and\r\n",
      "                        scaling factors between larger and smaller dataset to\r\n",
      "                        calculate statistics measurements. If you plan to use\r\n",
      "                        the signal output in bedGraph to call peaks using\r\n",
      "                        bdgcmp and bdgpeakcall, you shouldn't use this option\r\n",
      "                        because you will end up with different results.\r\n",
      "                        However, this option is recommended for displaying\r\n",
      "                        normalized pileup tracks across many datasets. Require\r\n",
      "                        -B to be set. Default: False\r\n",
      "\r\n",
      "Shifting model arguments:\r\n",
      "  --nomodel             Whether or not to build the shifting model. If True,\r\n",
      "                        MACS will not build model. by default it means\r\n",
      "                        shifting size = 100, try to set extsize to change it.\r\n",
      "                        It's highly recommended that while you have many\r\n",
      "                        datasets to process and you plan to compare different\r\n",
      "                        conditions, aka differential calling, use both\r\n",
      "                        'nomodel' and 'extsize' to make signal files from\r\n",
      "                        different datasets comparable. DEFAULT: False\r\n",
      "  --shift SHIFT         (NOT the legacy --shiftsize option!) The arbitrary\r\n",
      "                        shift in bp. Use discretion while setting it other\r\n",
      "                        than default value. When NOMODEL is set, MACS will use\r\n",
      "                        this value to move cutting ends (5') towards 5'->3'\r\n",
      "                        direction then apply EXTSIZE to extend them to\r\n",
      "                        fragments. When this value is negative, ends will be\r\n",
      "                        moved toward 3'->5' direction. Recommended to keep it\r\n",
      "                        as default 0 for ChIP-Seq datasets, or -1 * half of\r\n",
      "                        EXTSIZE together with EXTSIZE option for detecting\r\n",
      "                        enriched cutting loci such as certain DNAseI-Seq\r\n",
      "                        datasets. Note, you can't set values other than 0 if\r\n",
      "                        format is BAMPE or BEDPE for paired-end data. DEFAULT:\r\n",
      "                        0.\r\n",
      "  --extsize EXTSIZE     The arbitrary extension size in bp. When nomodel is\r\n",
      "                        true, MACS will use this value as fragment size to\r\n",
      "                        extend each read towards 3' end, then pile them up.\r\n",
      "                        It's exactly twice the number of obsolete SHIFTSIZE.\r\n",
      "                        In previous language, each read is moved 5'->3'\r\n",
      "                        direction to middle of fragment by 1/2 d, then\r\n",
      "                        extended to both direction with 1/2 d. This is\r\n",
      "                        equivalent to say each read is extended towards 5'->3'\r\n",
      "                        into a d size fragment. DEFAULT: 200. EXTSIZE and\r\n",
      "                        SHIFT can be combined when necessary. Check SHIFT\r\n",
      "                        option.\r\n",
      "  --bw BW               Band width for picking regions to compute fragment\r\n",
      "                        size. This value is only used while building the\r\n",
      "                        shifting model. Tweaking this is not recommended.\r\n",
      "                        DEFAULT: 300\r\n",
      "  -m MFOLD MFOLD, --mfold MFOLD MFOLD\r\n",
      "                        Select the regions within MFOLD range of high-\r\n",
      "                        confidence enrichment ratio against background to\r\n",
      "                        build model. Fold-enrichment in regions must be lower\r\n",
      "                        than upper limit, and higher than the lower limit. Use\r\n",
      "                        as \"-m 10 30\". This setting is only used while\r\n",
      "                        building the shifting model. Tweaking it is not\r\n",
      "                        recommended. DEFAULT:5 50\r\n",
      "  --fix-bimodal         Whether turn on the auto pair model process. If set,\r\n",
      "                        when MACS failed to build paired model, it will use\r\n",
      "                        the nomodel settings, the --exsize parameter to extend\r\n",
      "                        each tags towards 3' direction. Not to use this\r\n",
      "                        automate fixation is a default behavior now. DEFAULT:\r\n",
      "                        False\r\n",
      "\r\n",
      "Peak calling arguments:\r\n",
      "  -q QVALUE, --qvalue QVALUE\r\n",
      "                        Minimum FDR (q-value) cutoff for peak detection.\r\n",
      "                        DEFAULT: 0.05. -q, and -p are mutually exclusive.\r\n",
      "  -p PVALUE, --pvalue PVALUE\r\n",
      "                        Pvalue cutoff for peak detection. DEFAULT: not set.\r\n",
      "                        -q, and -p are mutually exclusive. If pvalue cutoff is\r\n",
      "                        set, qvalue will not be calculated and reported as -1\r\n",
      "                        in the final .xls file.\r\n",
      "  --scale-to {large,small}\r\n",
      "                        When set to 'small', scale the larger sample up to the\r\n",
      "                        smaller sample. When set to 'larger', scale the\r\n",
      "                        smaller sample up to the bigger sample. By default,\r\n",
      "                        scale to 'small'. This option replaces the obsolete\r\n",
      "                        '--to-large' option. The default behavior is\r\n",
      "                        recommended since it will lead to less significant\r\n",
      "                        p/q-values in general but more specific results. Keep\r\n",
      "                        in mind that scaling down will influence control/input\r\n",
      "                        sample more. DEFAULT: 'small', the choice is either\r\n",
      "                        'small' or 'large'.\r\n",
      "  --ratio RATIO         When set, use a custom scaling ratio of ChIP/control\r\n",
      "                        (e.g. calculated using NCIS) for linear scaling.\r\n",
      "                        DEFAULT: ingore\r\n",
      "  --down-sample         When set, random sampling method will scale down the\r\n",
      "                        bigger sample. By default, MACS uses linear scaling.\r\n",
      "                        Warning: This option will make your result unstable\r\n",
      "                        and irreproducible since each time, random reads would\r\n",
      "                        be selected. Consider to use 'randsample' script\r\n",
      "                        instead. <not implmented>If used together with --SPMR,\r\n",
      "                        1 million unique reads will be randomly picked.</not\r\n",
      "                        implemented> Caution: due to the implementation, the\r\n",
      "                        final number of selected reads may not be as you\r\n",
      "                        expected! DEFAULT: False\r\n",
      "  --seed SEED           Set the random seed while down sampling data. Must be\r\n",
      "                        a non-negative integer in order to be effective.\r\n",
      "                        DEFAULT: not set\r\n",
      "  --tempdir TEMPDIR     Optional directory to store temp files. DEFAULT:\r\n",
      "                        /var/folders/cs/w9tb85bx3qzdwmzvzp2q6hp00000gq/T\r\n",
      "  --nolambda            If True, MACS will use fixed background lambda as\r\n",
      "                        local lambda for every peak region. Normally, MACS\r\n",
      "                        calculates a dynamic local lambda to reflect the local\r\n",
      "                        bias due to the potential chromatin accessibility.\r\n",
      "  --slocal SMALLLOCAL   The small nearby region in basepairs to calculate\r\n",
      "                        dynamic lambda. This is used to capture the bias near\r\n",
      "                        the peak summit region. Invalid if there is no control\r\n",
      "                        data. If you set this to 0, MACS will skip slocal\r\n",
      "                        lambda calculation. *Note* that MACS will always\r\n",
      "                        perform a d-size local lambda calculation while the\r\n",
      "                        control data is available. The final local bias would\r\n",
      "                        be the maximum of the lambda value from d, slocal, and\r\n",
      "                        llocal size windows. While control is not available, d\r\n",
      "                        and slocal lambda won't be considered. DEFAULT: 1000\r\n",
      "  --llocal LARGELOCAL   The large nearby region in basepairs to calculate\r\n",
      "                        dynamic lambda. This is used to capture the surround\r\n",
      "                        bias. If you set this to 0, MACS will skip llocal\r\n",
      "                        lambda calculation. *Note* that MACS will always\r\n",
      "                        perform a d-size local lambda calculation while the\r\n",
      "                        control data is available. The final local bias would\r\n",
      "                        be the maximum of the lambda value from d, slocal, and\r\n",
      "                        llocal size windows. While control is not available, d\r\n",
      "                        and slocal lambda won't be considered. DEFAULT: 10000.\r\n",
      "  --max-gap MAXGAP      Maximum gap between significant sites to cluster them\r\n",
      "                        together. The DEFAULT value is the detected read\r\n",
      "                        length/tag size.\r\n",
      "  --min-length MINLEN   Minimum length of a peak. The DEFAULT value is the\r\n",
      "                        predicted fragment size d. Note, if you set a value\r\n",
      "                        smaller than the fragment size, it may have NO effect\r\n",
      "                        on the result. For BROAD peak calling, try to set a\r\n",
      "                        large value such as 500bps. You can also use\r\n",
      "                        '--cutoff-analysis' option with default setting, and\r\n",
      "                        check the column 'avelpeak' under different cutoff\r\n",
      "                        values to decide a reasonable minlen value.\r\n",
      "  --broad               If set, MACS will try to call broad peaks using the\r\n",
      "                        --broad-cutoff setting. Please tweak '--broad-cutoff'\r\n",
      "                        setting to control the peak calling behavior. At the\r\n",
      "                        meantime, either -q or -p cutoff will be used to\r\n",
      "                        define regions with 'stronger enrichment' inside of\r\n",
      "                        broad peaks. The maximum gap is expanded to 4 * MAXGAP\r\n",
      "                        (--max-gap parameter). As a result, MACS will output a\r\n",
      "                        'gappedPeak' and a 'broadPeak' file instead of\r\n",
      "                        'narrowPeak' file. Note, a broad peak will be reported\r\n",
      "                        even if there is no 'stronger enrichment' inside.\r\n",
      "                        DEFAULT: False\r\n",
      "  --broad-cutoff BROADCUTOFF\r\n",
      "                        Cutoff for broad region. This option is not available\r\n",
      "                        unless --broad is set. If -p is set, this is a pvalue\r\n",
      "                        cutoff, otherwise, it's a qvalue cutoff. Please note\r\n",
      "                        that in broad peakcalling mode, MACS2 uses this\r\n",
      "                        setting to control the overall peak calling behavior,\r\n",
      "                        then uses -q or -p setting to define regions inside\r\n",
      "                        broad region as 'stronger' enrichment. DEFAULT: 0.1\r\n",
      "  --cutoff-analysis     While set, MACS2 will analyze number or total length\r\n",
      "                        of peaks that can be called by different p-value\r\n",
      "                        cutoff then output a summary table to help user decide\r\n",
      "                        a better cutoff. The table will be saved in\r\n",
      "                        NAME_cutoff_analysis.txt file. Note, minlen and maxgap\r\n",
      "                        may affect the results. WARNING: May take ~30 folds\r\n",
      "                        longer time to finish. The result can be useful for\r\n",
      "                        users to decide a reasonable cutoff value. DEFAULT:\r\n",
      "                        False\r\n",
      "\r\n",
      "Post-processing options:\r\n",
      "  --call-summits        If set, MACS will use a more sophisticated signal\r\n",
      "                        processing approach to find subpeak summits in each\r\n",
      "                        enriched peak region. DEFAULT: False\r\n",
      "  --fe-cutoff FECUTOFF  When set, the value will be used to filter out peaks\r\n",
      "                        with low fold-enrichment. Note, MACS2 use 1.0 as\r\n",
      "                        pseudocount while calculating fold-enrichment.\r\n",
      "                        DEFAULT: 1.0\r\n",
      "\r\n",
      "Other options:\r\n",
      "  --buffer-size BUFFER_SIZE\r\n",
      "                        Buffer size for incrementally increasing internal\r\n",
      "                        array size to store reads alignment information. In\r\n",
      "                        most cases, you don't have to change this parameter.\r\n",
      "                        However, if there are large number of\r\n",
      "                        chromosomes/contigs/scaffolds in your alignment, it's\r\n",
      "                        recommended to specify a smaller buffer size in order\r\n",
      "                        to decrease memory usage (but it will take longer time\r\n",
      "                        to read alignment files). Minimum memory requested for\r\n",
      "                        reading an alignment file is about # of CHROMOSOME *\r\n",
      "                        BUFFER_SIZE * 2 Bytes. DEFAULT: 100000\r\n",
      "\r\n",
      "Obsolete options:\r\n",
      "  --to-large            Obsolete option. Please use '--scale-to large'\r\n",
      "                        instead.\r\n"
     ]
    }
   ],
   "source": [
    "! MACS2 callpeak -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.3.4.3 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\r\n",
      "Usage: \r\n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r> | --interleaved <i>} [-S <sam>]\r\n",
      "\r\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\r\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\r\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <r>        Files with unpaired reads.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <i>        Files with interleaved paired-end FASTQ reads\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <sam>      File for SAM output (default: stdout)\r\n",
      "\r\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\r\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\r\n",
      "\r\n",
      "Options (defaults in parentheses):\r\n",
      "\r\n",
      " Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  --tab5             query input files are TAB5 .tab5\r\n",
      "  --tab6             query input files are TAB6 .tab6\r\n",
      "  --qseq             query input files are in Illumina's qseq format\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -F k:<int>,i:<int> query input files are continuous FASTA where reads\r\n",
      "                     are substrings (k-mers) extracted from a FASTA file <s>\r\n",
      "                     and aligned at offsets 1, 1+i, 1+2i ... end of reference\r\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\r\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\r\n",
      "  --trim-to [3:|5:]<int> trim reads exceeding <int> bases from either 3' or 5' end\r\n",
      "                     If the read end is not specified then it defaults to 3 (0)\r\n",
      "  --phred33          qualities are Phred+33 (default)\r\n",
      "  --phred64          qualities are Phred+64\r\n",
      "  --int-quals        qualities encoded as space-delimited integers\r\n",
      "\r\n",
      " Presets:                 Same as:\r\n",
      "  For --end-to-end:\r\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\r\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      "  For --local:\r\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\r\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\r\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\r\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      " Alignment:\r\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\r\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\r\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\r\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\r\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\r\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\r\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\r\n",
      "  --nofw             do not align forward (original) version of read (off)\r\n",
      "  --norc             do not align reverse-complement version of read (off)\r\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\r\n",
      "                     scan for the optimal seeded alignments\r\n",
      "  --end-to-end       entire read must align; no clipping (on)\r\n",
      "   OR\r\n",
      "  --local            local alignment; ends might be soft clipped (off)\r\n",
      "\r\n",
      " Scoring:\r\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \r\n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\r\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\r\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\r\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\r\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\r\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\r\n",
      "\r\n",
      " Reporting:\r\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\r\n",
      "   OR\r\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\r\n",
      "   OR\r\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\r\n",
      "\r\n",
      " Effort:\r\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\r\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\r\n",
      "\r\n",
      " Paired-end:\r\n",
      "  -I/--minins <int>  minimum fragment length (0)\r\n",
      "  -X/--maxins <int>  maximum fragment length (500)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\r\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\r\n",
      "  --no-discordant    suppress discordant alignments for paired reads\r\n",
      "  --dovetail         concordant when mates extend past each other\r\n",
      "  --no-contain       not concordant when one mate alignment contains other\r\n",
      "  --no-overlap       not concordant when mates overlap at all\r\n",
      "\r\n",
      " Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  --un <path>        write unpaired reads that didn't align to <path>\r\n",
      "  --al <path>        write unpaired reads that aligned at least once to <path>\r\n",
      "  --un-conc <path>   write pairs that didn't align concordantly to <path>\r\n",
      "  --al-conc <path>   write pairs that aligned concordantly at least once to <path>\r\n",
      "    (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\r\n",
      "    --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\r\n",
      "  --quiet            print nothing to stderr except serious errors\r\n",
      "  --met-file <path>  send metrics to file at <path> (off)\r\n",
      "  --met-stderr       send metrics to stderr (off)\r\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\r\n",
      "  --no-unal          suppress SAM records for unaligned reads\r\n",
      "  --no-head          suppress header lines, i.e. lines starting with @\r\n",
      "  --no-sq            suppress @SQ header lines\r\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\r\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\r\n",
      "                     Note: @RG line only printed when --rg-id is set.\r\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\r\n",
      "  --sam-no-qname-trunc Suppress standard behavior of truncating readname at first whitespace \r\n",
      "                      at the expense of generating non-standard SAM.\r\n",
      "  --xeq              Use '='/'X', instead of 'M,' to specify matches/mismatches in SAM record.\r\n",
      "  --soft-clipped-unmapped-tlen Exclude soft-clipped bases when reporting TLEN\r\n",
      "\r\n",
      " Performance:\r\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\r\n",
      "  --reorder          force SAM output order to match order of input reads\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "\r\n",
      " Other:\r\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\r\n",
      "  --seed <int>       seed for random number generator (0)\r\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n"
     ]
    }
   ],
   "source": [
    "! bowtie2 -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's build the indexes with bowtie2  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 32\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  datatosend/reference.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:29\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:18\n",
      "bmax according to bmaxDivN setting: 89524538\n",
      "Using parameters --bmax 67143404 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 67143404 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:47\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:25\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:37\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 86 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 86 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 86 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 10, merged 40; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 7, merged 8; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 5, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 3, merged 5; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "Avg bucket size: 4.85557e+07 (target: 67143403)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 59\n",
      "  Reserving size (67143404) for bucket 1\n",
      "Getting block 2 of 59\n",
      "Getting block 3 of 59\n",
      "Getting block 5 of 59\n",
      "Getting block 4 of 59\n",
      "Getting block 7 of 59\n",
      "Getting block 6 of 59\n",
      "  Reserving size (67143404) for bucket 2\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Reserving size (67143404) for bucket 3\n",
      "  Reserving size (67143404) for bucket 5\n",
      "  Reserving size (67143404) for bucket 4\n",
      "  Reserving size (67143404) for bucket 7\n",
      "  Reserving size (67143404) for bucket 6\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 1: 10%\n",
      "  bucket 2: 10%\n",
      "  bucket 3: 10%\n",
      "  bucket 4: 10%\n",
      "  bucket 6: 10%\n",
      "  bucket 5: 10%\n",
      "  bucket 7: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 2: 20%\n",
      "  bucket 3: 20%\n",
      "  bucket 4: 20%\n",
      "  bucket 6: 20%\n",
      "  bucket 5: 20%\n",
      "  bucket 7: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 2: 30%\n",
      "  bucket 3: 30%\n",
      "  bucket 4: 30%\n",
      "  bucket 6: 30%\n",
      "  bucket 5: 30%\n",
      "  bucket 7: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 2: 40%\n",
      "  bucket 3: 40%\n",
      "  bucket 4: 40%\n",
      "  bucket 6: 40%\n",
      "  bucket 5: 40%\n",
      "  bucket 7: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 2: 50%\n",
      "  bucket 3: 50%\n",
      "  bucket 4: 50%\n",
      "  bucket 6: 50%\n",
      "  bucket 5: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 7: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 3: 60%\n",
      "  bucket 4: 60%\n",
      "  bucket 6: 60%\n",
      "  bucket 5: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 7: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 3: 70%\n",
      "  bucket 4: 70%\n",
      "  bucket 6: 70%\n",
      "  bucket 5: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 2: 80%\n",
      "  bucket 7: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 4: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 6: 80%\n",
      "  bucket 5: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 7: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 4: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 50254554 for bucket 1\n",
      "  (Using difference cover)\n",
      "  bucket 6: 90%\n",
      "  bucket 5: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 29517233 for bucket 2\n",
      "  (Using difference cover)\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 57546622 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 7: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 45026876 for bucket 4\n",
      "  (Using difference cover)\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 35162081 for bucket 6\n",
      "  (Using difference cover)\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 49485782 for bucket 5\n",
      "  (Using difference cover)\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 66567427 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:48\n",
      "Returning block of 29517234 for bucket 2\n",
      "Getting block 8 of 59\n",
      "  Reserving size (67143404) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  Sorting block time: 00:00:53\n",
      "Returning block of 35162082 for bucket 6\n",
      "Getting block 9 of 59\n",
      "  Reserving size (67143404) for bucket 9\n",
      "  Calculating Z arrays for bucket 9\n",
      "  Entering block accumulator loop for bucket 9:\n",
      "  bucket 8: 20%\n",
      "  bucket 9: 10%\n",
      "  Sorting block time: 00:01:05\n",
      "Returning block of 45026877 for bucket 4\n",
      "  bucket 8: 30%\n",
      "Getting block 10 of 59\n",
      "  Reserving size (67143404) for bucket 10\n",
      "  Calculating Z arrays for bucket 10\n",
      "  Entering block accumulator loop for bucket 10:\n",
      "  Sorting block time: 00:01:15\n",
      "Returning block of 50254555 for bucket 1\n",
      "  Sorting block time: 00:01:09\n",
      "Returning block of 49485783 for bucket 5\n",
      "Getting block 11 of 59\n",
      "  Reserving size (67143404) for bucket 11\n",
      "  Calculating Z arrays for bucket 11\n",
      "  Entering block accumulator loop for bucket 11:\n",
      "  bucket 9: 20%\n",
      "Getting block 12 of 59\n",
      "  Reserving size (67143404) for bucket 12\n",
      "  Calculating Z arrays for bucket 12\n",
      "  Entering block accumulator loop for bucket 12:\n",
      "  bucket 8: 40%\n",
      "  bucket 10: 10%\n",
      "  Sorting block time: 00:01:20\n",
      "Returning block of 57546623 for bucket 3\n",
      "  bucket 9: 30%\n",
      "  bucket 11: 10%\n",
      "  bucket 12: 10%\n",
      "Getting block 13 of 59\n",
      "  Reserving size (67143404) for bucket 13\n",
      "  Calculating Z arrays for bucket 13\n",
      "  Entering block accumulator loop for bucket 13:\n",
      "  bucket 8: 50%\n",
      "  bucket 10: 20%\n",
      "  bucket 9: 40%\n",
      "  bucket 11: 20%\n",
      "  bucket 12: 20%\n",
      "  bucket 13: 10%\n",
      "  bucket 8: 60%\n",
      "  bucket 10: 30%\n",
      "  Sorting block time: 00:01:26\n",
      "Returning block of 66567428 for bucket 7\n",
      "  bucket 9: 50%\n",
      "  bucket 11: 30%\n",
      "Getting block 14 of 59\n",
      "  Reserving size (67143404) for bucket 14\n",
      "  Calculating Z arrays for bucket 14\n",
      "  Entering block accumulator loop for bucket 14:\n",
      "  bucket 8: 70%\n",
      "  bucket 12: 30%\n",
      "  bucket 13: 20%\n",
      "  bucket 10: 40%\n",
      "  bucket 9: 60%\n",
      "  bucket 11: 40%\n",
      "  bucket 14: 10%\n",
      "  bucket 8: 80%\n",
      "  bucket 12: 40%\n",
      "  bucket 13: 30%\n",
      "  bucket 10: 50%\n",
      "  bucket 9: 70%\n",
      "  bucket 11: 50%\n",
      "  bucket 8: 90%\n",
      "  bucket 14: 20%\n",
      "  bucket 12: 50%\n",
      "  bucket 13: 40%\n",
      "  bucket 10: 60%\n",
      "  bucket 9: 80%\n",
      "  bucket 11: 60%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 25793610 for bucket 8\n",
      "  (Using difference cover)\n",
      "  bucket 14: 30%\n",
      "  bucket 12: 60%\n",
      "  bucket 10: 70%\n",
      "  bucket 13: 50%\n",
      "  bucket 9: 90%\n",
      "  bucket 11: 70%\n",
      "  bucket 14: 40%\n",
      "  bucket 10: 80%\n",
      "  bucket 12: 70%\n",
      "  bucket 13: 60%\n",
      "  bucket 9: 100%\n",
      "  Sorting block of length 58700327 for bucket 9\n",
      "  (Using difference cover)\n",
      "  bucket 11: 80%\n",
      "  bucket 10: 90%\n",
      "  bucket 14: 50%\n",
      "  bucket 12: 80%\n",
      "  bucket 13: 70%\n",
      "  bucket 11: 90%\n",
      "  bucket 10: 100%\n",
      "  Sorting block of length 19289144 for bucket 10\n",
      "  (Using difference cover)\n",
      "  bucket 14: 60%\n",
      "  bucket 12: 90%\n",
      "  bucket 13: 80%\n",
      "  Sorting block time: 00:00:29\n",
      "Returning block of 25793611 for bucket 8\n",
      "  bucket 11: 100%\n",
      "  Sorting block of length 64865745 for bucket 11\n",
      "  (Using difference cover)\n",
      "Getting block 15 of 59\n",
      "  Reserving size (67143404) for bucket 15\n",
      "  Calculating Z arrays for bucket 15\n",
      "  Entering block accumulator loop for bucket 15:\n",
      "  bucket 12: 100%\n",
      "  Sorting block of length 51208554 for bucket 12\n",
      "  (Using difference cover)\n",
      "  bucket 14: 70%\n",
      "  bucket 13: 90%\n",
      "  bucket 15: 10%\n",
      "  bucket 14: 80%\n",
      "  bucket 13: 100%\n",
      "  Sorting block of length 20885795 for bucket 13\n",
      "  (Using difference cover)\n",
      "  bucket 15: 20%\n",
      "  Sorting block time: 00:00:23\n",
      "Returning block of 19289145 for bucket 10\n",
      "Getting block 16 of 59\n",
      "  Reserving size (67143404) for bucket 16\n",
      "  Calculating Z arrays for bucket 16\n",
      "  Entering block accumulator loop for bucket 16:\n",
      "  bucket 14: 90%\n",
      "  bucket 15: 30%\n",
      "  bucket 16: 10%\n",
      "  bucket 14: 100%\n",
      "  Sorting block of length 55701547 for bucket 14\n",
      "  (Using difference cover)\n",
      "  bucket 15: 40%\n",
      "  bucket 16: 20%\n",
      "  Sorting block time: 00:00:27\n",
      "Returning block of 20885796 for bucket 13\n",
      "Getting block 17 of 59\n",
      "  Reserving size (67143404) for bucket 17\n",
      "  Calculating Z arrays for bucket 17\n",
      "  Entering block accumulator loop for bucket 17:\n",
      "  bucket 15: 50%\n",
      "  bucket 16: 30%\n",
      "  bucket 17: 10%\n",
      "  bucket 15: 60%\n",
      "  bucket 16: 40%\n",
      "  Sorting block time: 00:01:11\n",
      "Returning block of 58700328 for bucket 9\n",
      "  bucket 17: 20%\n",
      "  bucket 15: 70%\n",
      "Getting block 18 of 59\n",
      "  Reserving size (67143404) for bucket 18\n",
      "  Calculating Z arrays for bucket 18\n",
      "  Entering block accumulator loop for bucket 18:\n",
      "  bucket 16: 50%\n",
      "  bucket 15: 80%\n",
      "  bucket 17: 30%\n",
      "  bucket 18: 10%\n",
      "  bucket 16: 60%\n",
      "  Sorting block time: 00:01:04\n",
      "Returning block of 51208555 for bucket 12\n",
      "  bucket 15: 90%\n",
      "  bucket 17: 40%\n",
      "Getting block 19 of 59\n",
      "  Reserving size (67143404) for bucket 19\n",
      "  Calculating Z arrays for bucket 19\n",
      "  Entering block accumulator loop for bucket 19:\n",
      "  bucket 18: 20%\n",
      "  bucket 16: 70%\n",
      "  Sorting block time: 00:01:17\n",
      "Returning block of 64865746 for bucket 11\n",
      "  bucket 15: 100%\n",
      "  Sorting block of length 28466842 for bucket 15\n",
      "  (Using difference cover)\n",
      "  bucket 17: 50%\n",
      "Getting block 20 of 59\n",
      "  Reserving size (67143404) for bucket 20\n",
      "  Calculating Z arrays for bucket 20\n",
      "  Entering block accumulator loop for bucket 20:\n",
      "  bucket 19: 10%\n",
      "  bucket 18: 30%\n",
      "  bucket 16: 80%\n",
      "  bucket 17: 60%\n",
      "  bucket 20: 10%\n",
      "  bucket 19: 20%\n",
      "  bucket 18: 40%\n",
      "  bucket 16: 90%\n",
      "  bucket 17: 70%\n",
      "  Sorting block time: 00:01:07\n",
      "Returning block of 55701548 for bucket 14\n",
      "  bucket 20: 20%\n",
      "Getting block 21 of 59\n",
      "  Reserving size (67143404) for bucket 21\n",
      "  Calculating Z arrays for bucket 21\n",
      "  Entering block accumulator loop for bucket 21:\n",
      "  bucket 18: 50%\n",
      "  bucket 16: 100%\n",
      "  Sorting block of length 52583734 for bucket 16\n",
      "  (Using difference cover)\n",
      "  bucket 19: 30%\n",
      "  bucket 17: 80%\n",
      "  bucket 20: 30%\n",
      "  bucket 21: 10%\n",
      "  bucket 18: 60%\n",
      "  bucket 19: 40%\n",
      "  Sorting block time: 00:00:33\n",
      "Returning block of 28466843 for bucket 15\n",
      "  bucket 17: 90%\n",
      "Getting block 22 of 59\n",
      "  Reserving size (67143404) for bucket 22\n",
      "  Calculating Z arrays for bucket 22\n",
      "  Entering block accumulator loop for bucket 22:\n",
      "  bucket 20: 40%\n",
      "  bucket 21: 20%\n",
      "  bucket 18: 70%\n",
      "  bucket 19: 50%\n",
      "  bucket 17: 100%\n",
      "  Sorting block of length 26338819 for bucket 17\n",
      "  (Using difference cover)\n",
      "  bucket 22: 10%\n",
      "  bucket 20: 50%\n",
      "  bucket 21: 30%\n",
      "  bucket 18: 80%\n",
      "  bucket 19: 60%\n",
      "  bucket 22: 20%\n",
      "  bucket 20: 60%\n",
      "  bucket 21: 40%\n",
      "  bucket 18: 90%\n",
      "  bucket 19: 70%\n",
      "  bucket 22: 30%\n",
      "  bucket 20: 70%\n",
      "  bucket 21: 50%\n",
      "  bucket 18: 100%\n",
      "  Sorting block of length 66846756 for bucket 18\n",
      "  (Using difference cover)\n",
      "  bucket 19: 80%\n",
      "  bucket 22: 40%\n",
      "  bucket 20: 80%\n",
      "  Sorting block time: 00:00:31\n",
      "Returning block of 26338820 for bucket 17\n",
      "  bucket 21: 60%\n",
      "Getting block 23 of 59\n",
      "  Reserving size (67143404) for bucket 23\n",
      "  Calculating Z arrays for bucket 23\n",
      "  Entering block accumulator loop for bucket 23:\n",
      "  bucket 19: 90%\n",
      "  bucket 22: 50%\n",
      "  bucket 20: 90%\n",
      "  Sorting block time: 00:00:59\n",
      "Returning block of 52583735 for bucket 16\n",
      "  bucket 21: 70%\n",
      "  bucket 23: 10%\n",
      "Getting block 24 of 59\n",
      "  Reserving size (67143404) for bucket 24\n",
      "  Calculating Z arrays for bucket 24\n",
      "  Entering block accumulator loop for bucket 24:\n",
      "  bucket 22: 60%\n",
      "  bucket 19: 100%\n",
      "  Sorting block of length 60942175 for bucket 19\n",
      "  (Using difference cover)\n",
      "  bucket 20: 100%\n",
      "  Sorting block of length 62834064 for bucket 20\n",
      "  (Using difference cover)\n",
      "  bucket 23: 20%\n",
      "  bucket 21: 80%\n",
      "  bucket 24: 10%\n",
      "  bucket 22: 70%\n",
      "  bucket 23: 30%\n",
      "  bucket 21: 90%\n",
      "  bucket 24: 20%\n",
      "  bucket 22: 80%\n",
      "  bucket 23: 40%\n",
      "  bucket 21: 100%\n",
      "  Sorting block of length 57188208 for bucket 21\n",
      "  (Using difference cover)\n",
      "  bucket 24: 30%\n",
      "  bucket 22: 90%\n",
      "  bucket 23: 50%\n",
      "  bucket 24: 40%\n",
      "  bucket 23: 60%\n",
      "  bucket 22: 100%\n",
      "  Sorting block of length 89846872 for bucket 22\n",
      "  (Using difference cover)\n",
      "  bucket 24: 50%\n",
      "  bucket 23: 70%\n",
      "  bucket 24: 60%\n",
      "  bucket 23: 80%\n",
      "  bucket 24: 70%\n",
      "  bucket 23: 90%\n",
      "  Sorting block time: 00:01:20\n",
      "Returning block of 66846757 for bucket 18\n",
      "  bucket 24: 80%\n",
      "Getting block 25 of 59\n",
      "  Reserving size (67143404) for bucket 25\n",
      "  Calculating Z arrays for bucket 25\n",
      "  Entering block accumulator loop for bucket 25:\n",
      "  bucket 23: 100%\n",
      "  Sorting block of length 1433540 for bucket 23\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:03\n",
      "Returning block of 1433541 for bucket 23\n",
      "Getting block 26 of 59\n",
      "  Reserving size (67143404) for bucket 26\n",
      "  Calculating Z arrays for bucket 26\n",
      "  Entering block accumulator loop for bucket 26:\n",
      "  bucket 24: 90%\n",
      "  bucket 25: 10%\n",
      "  Sorting block time: 00:01:16\n",
      "Returning block of 60942176 for bucket 19\n",
      "Getting block 27 of 59\n",
      "  Reserving size (67143404) for bucket 27\n",
      "  Calculating Z arrays for bucket 27\n",
      "  Entering block accumulator loop for bucket 27:\n",
      "  bucket 26: 10%\n",
      "  Sorting block time: 00:01:19\n",
      "Returning block of 62834065 for bucket 20\n",
      "  bucket 24: 100%\n",
      "  Sorting block of length 25662163 for bucket 24\n",
      "  (Using difference cover)\n",
      "  bucket 25: 20%\n",
      "Getting block 28 of 59\n",
      "  Reserving size (67143404) for bucket 28\n",
      "  Calculating Z arrays for bucket 28\n",
      "  Entering block accumulator loop for bucket 28:\n",
      "  bucket 27: 10%\n",
      "  bucket 26: 20%\n",
      "  bucket 25: 30%\n",
      "  bucket 28: 10%\n",
      "  Sorting block time: 00:01:11\n",
      "Returning block of 57188209 for bucket 21\n",
      "Getting block 29 of 59\n",
      "  Reserving size (67143404) for bucket 29\n",
      "  Calculating Z arrays for bucket 29\n",
      "  Entering block accumulator loop for bucket 29:\n",
      "  bucket 27: 20%\n",
      "  bucket 26: 30%\n",
      "  bucket 25: 40%\n",
      "  bucket 28: 20%\n",
      "  bucket 29: 10%\n",
      "  bucket 27: 30%\n",
      "  bucket 26: 40%\n",
      "  bucket 25: 50%\n",
      "  bucket 28: 30%\n",
      "  Sorting block time: 00:00:29\n",
      "Returning block of 25662164 for bucket 24\n",
      "Getting block 30 of 59\n",
      "  Reserving size (67143404) for bucket 30\n",
      "  Calculating Z arrays for bucket 30\n",
      "  Entering block accumulator loop for bucket 30:\n",
      "  bucket 27: 40%\n",
      "  bucket 26: 50%\n",
      "  bucket 29: 20%\n",
      "  bucket 25: 60%\n",
      "  bucket 28: 40%\n",
      "  bucket 30: 10%\n",
      "  bucket 26: 60%\n",
      "  bucket 27: 50%\n",
      "  bucket 25: 70%\n",
      "  bucket 29: 30%\n",
      "  bucket 28: 50%\n",
      "  bucket 30: 20%\n",
      "  bucket 25: 80%\n",
      "  bucket 26: 70%\n",
      "  bucket 27: 60%\n",
      "  bucket 29: 40%\n",
      "  bucket 28: 60%\n",
      "  Sorting block time: 00:01:40\n",
      "Returning block of 89846873 for bucket 22\n",
      "  bucket 30: 30%\n",
      "  bucket 25: 90%\n",
      "  bucket 26: 80%\n",
      "  bucket 27: 70%\n",
      "  bucket 29: 50%\n",
      "Getting block 31 of 59\n",
      "  Reserving size (67143404) for bucket 31\n",
      "  Calculating Z arrays for bucket 31\n",
      "  Entering block accumulator loop for bucket 31:\n",
      "  bucket 28: 70%\n",
      "  bucket 25: 100%\n",
      "  Sorting block of length 48429724 for bucket 25\n",
      "  (Using difference cover)\n",
      "  bucket 30: 40%\n",
      "  bucket 26: 90%\n",
      "  bucket 27: 80%\n",
      "  bucket 29: 60%\n",
      "  bucket 31: 10%\n",
      "  bucket 28: 80%\n",
      "  bucket 30: 50%\n",
      "  bucket 26: 100%\n",
      "  Sorting block of length 56938618 for bucket 26\n",
      "  (Using difference cover)\n",
      "  bucket 27: 90%\n",
      "  bucket 29: 70%\n",
      "  bucket 28: 90%\n",
      "  bucket 31: 20%\n",
      "  bucket 27: 100%\n",
      "  Sorting block of length 66169006 for bucket 27\n",
      "  (Using difference cover)\n",
      "  bucket 30: 60%\n",
      "  bucket 29: 80%\n",
      "  bucket 28: 100%\n",
      "  Sorting block of length 18322031 for bucket 28\n",
      "  (Using difference cover)\n",
      "  bucket 31: 30%\n",
      "  bucket 30: 70%\n",
      "  bucket 29: 90%\n",
      "  bucket 31: 40%\n",
      "  bucket 30: 80%\n",
      "  bucket 29: 100%\n",
      "  Sorting block of length 51014201 for bucket 29\n",
      "  (Using difference cover)\n",
      "  bucket 31: 50%\n",
      "  bucket 30: 90%\n",
      "  Sorting block time: 00:00:24\n",
      "Returning block of 18322032 for bucket 28\n",
      "Getting block 32 of 59\n",
      "  Reserving size (67143404) for bucket 32\n",
      "  Calculating Z arrays for bucket 32\n",
      "  Entering block accumulator loop for bucket 32:\n",
      "  bucket 31: 60%\n",
      "  bucket 30: 100%\n",
      "  Sorting block of length 34871920 for bucket 30\n",
      "  (Using difference cover)\n",
      "  bucket 32: 10%\n",
      "  Sorting block time: 00:01:01\n",
      "Returning block of 48429725 for bucket 25\n",
      "Getting block 33 of 59\n",
      "  Reserving size (67143404) for bucket 33\n",
      "  Calculating Z arrays for bucket 33\n",
      "  Entering block accumulator loop for bucket 33:\n",
      "  bucket 31: 70%\n",
      "  bucket 32: 20%\n",
      "  bucket 33: 10%\n",
      "  bucket 32: 30%\n",
      "  bucket 31: 80%\n",
      "  bucket 33: 20%\n",
      "  bucket 32: 40%\n",
      "  Sorting block time: 00:01:19\n",
      "Returning block of 56938619 for bucket 26\n",
      "  bucket 31: 90%\n",
      "Getting block 34 of 59\n",
      "  Reserving size (67143404) for bucket 34\n",
      "  Calculating Z arrays for bucket 34\n",
      "  Entering block accumulator loop for bucket 34:\n",
      "  bucket 33: 30%\n",
      "  bucket 32: 50%\n",
      "  bucket 31: 100%\n",
      "  Sorting block of length 66343918 for bucket 31\n",
      "  (Using difference cover)\n",
      "  bucket 34: 10%\n",
      "  bucket 33: 40%\n",
      "  bucket 32: 60%\n",
      "  Sorting block time: 00:00:49\n",
      "Returning block of 34871921 for bucket 30\n",
      "Getting block 35 of 59\n",
      "  Reserving size (67143404) for bucket 35\n",
      "  Calculating Z arrays for bucket 35\n",
      "  Entering block accumulator loop for bucket 35:\n",
      "  Sorting block time: 00:01:30\n",
      "Returning block of 66169007 for bucket 27\n",
      "  Sorting block time: 00:01:11\n",
      "Returning block of 51014202 for bucket 29\n",
      "  bucket 34: 20%\n",
      "  bucket 33: 50%\n",
      "Getting block 36 of 59\n",
      "  Reserving size (67143404) for bucket 36\n",
      "  Calculating Z arrays for bucket 36\n",
      "  Entering block accumulator loop for bucket 36:\n",
      "Getting block 37 of 59\n",
      "  Reserving size (67143404) for bucket 37\n",
      "  Calculating Z arrays for bucket 37\n",
      "  Entering block accumulator loop for bucket 37:\n",
      "  bucket 32: 70%\n",
      "  bucket 35: 10%\n",
      "  bucket 34: 30%\n",
      "  bucket 33: 60%\n",
      "  bucket 36: 10%\n",
      "  bucket 37: 10%\n",
      "  bucket 32: 80%\n",
      "  bucket 35: 20%\n",
      "  bucket 34: 40%\n",
      "  bucket 33: 70%\n",
      "  bucket 36: 20%\n",
      "  bucket 37: 20%\n",
      "  bucket 32: 90%\n",
      "  bucket 35: 30%\n",
      "  bucket 34: 50%\n",
      "  bucket 33: 80%\n",
      "  bucket 36: 30%\n",
      "  bucket 37: 30%\n",
      "  bucket 32: 100%\n",
      "  Sorting block of length 25464625 for bucket 32\n",
      "  (Using difference cover)\n",
      "  bucket 35: 40%\n",
      "  bucket 34: 60%\n",
      "  bucket 33: 90%\n",
      "  bucket 36: 40%\n",
      "  bucket 37: 40%\n",
      "  bucket 35: 50%\n",
      "  bucket 34: 70%\n",
      "  bucket 33: 100%\n",
      "  Sorting block of length 64805058 for bucket 33\n",
      "  (Using difference cover)\n",
      "  bucket 36: 50%\n",
      "  bucket 37: 50%\n",
      "  bucket 35: 60%\n",
      "  bucket 34: 80%\n",
      "  bucket 36: 60%\n",
      "  bucket 37: 60%\n",
      "  Sorting block time: 00:00:37\n",
      "Returning block of 25464626 for bucket 32\n",
      "Getting block 38 of 59\n",
      "  Reserving size (67143404) for bucket 38\n",
      "  Calculating Z arrays for bucket 38\n",
      "  Entering block accumulator loop for bucket 38:\n",
      "  bucket 35: 70%\n",
      "  bucket 34: 90%\n",
      "  bucket 36: 70%\n",
      "  Sorting block time: 00:01:25\n",
      "Returning block of 66343919 for bucket 31\n",
      "  bucket 37: 70%\n",
      "Getting block 39 of 59\n",
      "  Reserving size (67143404) for bucket 39\n",
      "  Calculating Z arrays for bucket 39\n",
      "  Entering block accumulator loop for bucket 39:\n",
      "  bucket 38: 10%\n",
      "  bucket 34: 100%\n",
      "  Sorting block of length 65567451 for bucket 34\n",
      "  (Using difference cover)\n",
      "  bucket 35: 80%\n",
      "  bucket 36: 80%\n",
      "  bucket 37: 80%\n",
      "  bucket 39: 10%\n",
      "  bucket 38: 20%\n",
      "  bucket 35: 90%\n",
      "  bucket 36: 90%\n",
      "  bucket 37: 90%\n",
      "  bucket 39: 20%\n",
      "  bucket 38: 30%\n",
      "  bucket 35: 100%\n",
      "  Sorting block of length 47856161 for bucket 35\n",
      "  (Using difference cover)\n",
      "  bucket 36: 100%\n",
      "  Sorting block of length 32750104 for bucket 36\n",
      "  (Using difference cover)\n",
      "  bucket 37: 100%\n",
      "  Sorting block of length 59494867 for bucket 37\n",
      "  (Using difference cover)\n",
      "  bucket 39: 30%\n",
      "  bucket 38: 40%\n",
      "  bucket 39: 40%\n",
      "  bucket 38: 50%\n",
      "  bucket 39: 50%\n",
      "  bucket 38: 60%\n",
      "  bucket 39: 60%\n",
      "  Sorting block time: 00:01:26\n",
      "Returning block of 64805059 for bucket 33\n",
      "  bucket 38: 70%\n",
      "Getting block 40 of 59\n",
      "  Reserving size (67143404) for bucket 40\n",
      "  Calculating Z arrays for bucket 40\n",
      "  Entering block accumulator loop for bucket 40:\n",
      "  Sorting block time: 00:00:41\n",
      "Returning block of 32750105 for bucket 36\n",
      "  bucket 39: 70%\n",
      "Getting block 41 of 59\n",
      "  Reserving size (67143404) for bucket 41\n",
      "  Calculating Z arrays for bucket 41\n",
      "  Entering block accumulator loop for bucket 41:\n",
      "  bucket 38: 80%\n",
      "  bucket 40: 10%\n",
      "  bucket 39: 80%\n",
      "  bucket 41: 10%\n",
      "  bucket 38: 90%\n",
      "  Sorting block time: 00:00:54\n",
      "Returning block of 47856162 for bucket 35\n",
      "  bucket 40: 20%\n",
      "  Sorting block time: 00:01:15\n",
      "Returning block of 65567452 for bucket 34\n",
      "Getting block 42 of 59\n",
      "  Reserving size (67143404) for bucket 42\n",
      "  Calculating Z arrays for bucket 42\n",
      "  Entering block accumulator loop for bucket 42:\n",
      "Getting block 43 of 59\n",
      "  Reserving size (67143404) for bucket 43\n",
      "  Calculating Z arrays for bucket 43\n",
      "  Entering block accumulator loop for bucket 43:\n",
      "  bucket 39: 90%\n",
      "  bucket 41: 20%\n",
      "  bucket 38: 100%\n",
      "  Sorting block of length 55664974 for bucket 38\n",
      "  (Using difference cover)\n",
      "  bucket 40: 30%\n",
      "  bucket 42: 10%\n",
      "  bucket 43: 10%\n",
      "  bucket 39: 100%\n",
      "  Sorting block of length 64804668 for bucket 39\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:06\n",
      "Returning block of 59494868 for bucket 37\n",
      "  bucket 41: 30%\n",
      "Getting block 44 of 59\n",
      "  Reserving size (67143404) for bucket 44\n",
      "  Calculating Z arrays for bucket 44\n",
      "  Entering block accumulator loop for bucket 44:\n",
      "  bucket 40: 40%\n",
      "  bucket 42: 20%\n",
      "  bucket 43: 20%\n",
      "  bucket 41: 40%\n",
      "  bucket 44: 10%\n",
      "  bucket 40: 50%\n",
      "  bucket 43: 30%\n",
      "  bucket 42: 30%\n",
      "  bucket 41: 50%\n",
      "  bucket 44: 20%\n",
      "  bucket 40: 60%\n",
      "  bucket 43: 40%\n",
      "  bucket 42: 40%\n",
      "  bucket 41: 60%\n",
      "  bucket 44: 30%\n",
      "  bucket 40: 70%\n",
      "  bucket 43: 50%\n",
      "  bucket 42: 50%\n",
      "  bucket 41: 70%\n",
      "  bucket 44: 40%\n",
      "  bucket 40: 80%\n",
      "  bucket 43: 60%\n",
      "  bucket 42: 60%\n",
      "  bucket 41: 80%\n",
      "  bucket 44: 50%\n",
      "  bucket 40: 90%\n",
      "  Sorting block time: 00:00:56\n",
      "Returning block of 55664975 for bucket 38\n",
      "  bucket 43: 70%\n",
      "Getting block 45 of 59\n",
      "  Reserving size (67143404) for bucket 45\n",
      "  Calculating Z arrays for bucket 45\n",
      "  Entering block accumulator loop for bucket 45:\n",
      "  bucket 42: 70%\n",
      "  bucket 41: 90%\n",
      "  bucket 44: 60%\n",
      "  bucket 40: 100%\n",
      "  Sorting block of length 51696590 for bucket 40\n",
      "  (Using difference cover)\n",
      "  bucket 43: 80%\n",
      "  bucket 45: 10%\n",
      "  Sorting block time: 00:01:02\n",
      "Returning block of 64804669 for bucket 39\n",
      "Getting block 46 of 59\n",
      "  Reserving size (67143404) for bucket 46\n",
      "  Calculating Z arrays for bucket 46\n",
      "  Entering block accumulator loop for bucket 46:\n",
      "  bucket 42: 80%\n",
      "  bucket 44: 70%\n",
      "  bucket 41: 100%\n",
      "  Sorting block of length 48796601 for bucket 41\n",
      "  (Using difference cover)\n",
      "  bucket 43: 90%\n",
      "  bucket 45: 20%\n",
      "  bucket 44: 80%\n",
      "  bucket 46: 10%\n",
      "  bucket 42: 90%\n",
      "  bucket 43: 100%\n",
      "  Sorting block of length 58368493 for bucket 43\n",
      "  (Using difference cover)\n",
      "  bucket 45: 30%\n",
      "  bucket 44: 90%\n",
      "  bucket 46: 20%\n",
      "  bucket 42: 100%\n",
      "  Sorting block of length 59070888 for bucket 42\n",
      "  (Using difference cover)\n",
      "  bucket 45: 40%\n",
      "  bucket 44: 100%\n",
      "  Sorting block of length 38708026 for bucket 44\n",
      "  (Using difference cover)\n",
      "  bucket 46: 30%\n",
      "  bucket 45: 50%\n",
      "  bucket 46: 40%\n",
      "  bucket 45: 60%\n",
      "  bucket 46: 50%\n",
      "  bucket 45: 70%\n",
      "  Sorting block time: 00:01:09\n",
      "Returning block of 51696591 for bucket 40\n",
      "Getting block 47 of 59\n",
      "  Reserving size (67143404) for bucket 47\n",
      "  Calculating Z arrays for bucket 47\n",
      "  Entering block accumulator loop for bucket 47:\n",
      "  bucket 46: 60%\n",
      "  bucket 45: 80%\n",
      "  Sorting block time: 00:01:08\n",
      "Returning block of 48796602 for bucket 41\n",
      "Getting block 48 of 59\n",
      "  Reserving size (67143404) for bucket 48\n",
      "  Calculating Z arrays for bucket 48\n",
      "  Entering block accumulator loop for bucket 48:\n",
      "  bucket 47: 10%\n",
      "  bucket 46: 70%\n",
      "  bucket 45: 90%\n",
      "  Sorting block time: 00:00:45\n",
      "Returning block of 38708027 for bucket 44\n",
      "  bucket 48: 10%\n",
      "Getting block 49 of 59\n",
      "  Reserving size (67143404) for bucket 49\n",
      "  Calculating Z arrays for bucket 49\n",
      "  Entering block accumulator loop for bucket 49:\n",
      "  bucket 47: 20%\n",
      "  bucket 46: 80%\n",
      "  Sorting block time: 00:01:10\n",
      "Returning block of 58368494 for bucket 43\n",
      "  bucket 45: 100%\n",
      "  Sorting block of length 55063650 for bucket 45\n",
      "  (Using difference cover)\n",
      "Getting block 50 of 59\n",
      "  Reserving size (67143404) for bucket 50\n",
      "  Calculating Z arrays for bucket 50\n",
      "  Entering block accumulator loop for bucket 50:\n",
      "  bucket 48: 20%\n",
      "  bucket 49: 10%\n",
      "  bucket 47: 30%\n",
      "  bucket 46: 90%\n",
      "  Sorting block time: 00:01:10\n",
      "Returning block of 59070889 for bucket 42\n",
      "  bucket 50: 10%\n",
      "Getting block 51 of 59\n",
      "  Reserving size (67143404) for bucket 51\n",
      "  Calculating Z arrays for bucket 51\n",
      "  Entering block accumulator loop for bucket 51:\n",
      "  bucket 48: 30%\n",
      "  bucket 49: 20%\n",
      "  bucket 47: 40%\n",
      "  bucket 46: 100%\n",
      "  Sorting block of length 57690850 for bucket 46\n",
      "  (Using difference cover)\n",
      "  bucket 50: 20%\n",
      "  bucket 51: 10%\n",
      "  bucket 48: 40%\n",
      "  bucket 49: 30%\n",
      "  bucket 47: 50%\n",
      "  bucket 50: 30%\n",
      "  bucket 51: 20%\n",
      "  bucket 48: 50%\n",
      "  bucket 49: 40%\n",
      "  bucket 47: 60%\n",
      "  bucket 50: 40%\n",
      "  bucket 48: 60%\n",
      "  bucket 51: 30%\n",
      "  bucket 49: 50%\n",
      "  bucket 47: 70%\n",
      "  bucket 50: 50%\n",
      "  bucket 48: 70%\n",
      "  bucket 51: 40%\n",
      "  bucket 49: 60%\n",
      "  Sorting block time: 00:00:52\n",
      "Returning block of 55063651 for bucket 45\n",
      "  bucket 47: 80%\n",
      "Getting block 52 of 59\n",
      "  Reserving size (67143404) for bucket 52\n",
      "  Calculating Z arrays for bucket 52\n",
      "  Entering block accumulator loop for bucket 52:\n",
      "  bucket 50: 60%\n",
      "  bucket 48: 80%\n",
      "  bucket 51: 50%\n",
      "  bucket 49: 70%\n",
      "  bucket 47: 90%\n",
      "  bucket 52: 10%\n",
      "  bucket 50: 70%\n",
      "  bucket 48: 90%\n",
      "  bucket 51: 60%\n",
      "  bucket 49: 80%\n",
      "  Sorting block time: 00:00:55\n",
      "Returning block of 57690851 for bucket 46\n",
      "  bucket 47: 100%\n",
      "  Sorting block of length 57757095 for bucket 47\n",
      "  (Using difference cover)\n",
      "Getting block 53 of 59\n",
      "  Reserving size (67143404) for bucket 53\n",
      "  Calculating Z arrays for bucket 53\n",
      "  Entering block accumulator loop for bucket 53:\n",
      "  bucket 52: 20%\n",
      "  bucket 50: 80%\n",
      "  bucket 48: 100%\n",
      "  Sorting block of length 24383706 for bucket 48\n",
      "  (Using difference cover)\n",
      "  bucket 51: 70%\n",
      "  bucket 49: 90%\n",
      "  bucket 53: 10%\n",
      "  bucket 52: 30%\n",
      "  bucket 50: 90%\n",
      "  bucket 51: 80%\n",
      "  bucket 49: 100%\n",
      "  Sorting block of length 63037284 for bucket 49\n",
      "  (Using difference cover)\n",
      "  bucket 53: 20%\n",
      "  bucket 52: 40%\n",
      "  bucket 50: 100%\n",
      "  Sorting block of length 15726418 for bucket 50\n",
      "  (Using difference cover)\n",
      "  bucket 51: 90%\n",
      "  Sorting block time: 00:00:24\n",
      "Returning block of 24383707 for bucket 48\n",
      "  bucket 53: 30%\n",
      "  bucket 52: 50%\n",
      "Getting block 54 of 59\n",
      "  Reserving size (67143404) for bucket 54\n",
      "  Calculating Z arrays for bucket 54\n",
      "  Entering block accumulator loop for bucket 54:\n",
      "  bucket 51: 100%\n",
      "  Sorting block of length 65201027 for bucket 51\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:15\n",
      "Returning block of 15726419 for bucket 50\n",
      "  bucket 53: 40%\n",
      "  bucket 52: 60%\n",
      "Getting block 55 of 59\n",
      "  Reserving size (67143404) for bucket 55\n",
      "  Calculating Z arrays for bucket 55\n",
      "  Entering block accumulator loop for bucket 55:\n",
      "  bucket 54: 10%\n",
      "  bucket 53: 50%\n",
      "  bucket 52: 70%\n",
      "  bucket 55: 10%\n",
      "  bucket 54: 20%\n",
      "  bucket 55: 20%\n",
      "  bucket 53: 60%\n",
      "  Sorting block time: 00:00:55\n",
      "Returning block of 57757096 for bucket 47\n",
      "  bucket 52: 80%\n",
      "  bucket 54: 30%\n",
      "Getting block 56 of 59\n",
      "  Reserving size (67143404) for bucket 56\n",
      "  Calculating Z arrays for bucket 56\n",
      "  Entering block accumulator loop for bucket 56:\n",
      "  bucket 55: 30%\n",
      "  bucket 53: 70%\n",
      "  bucket 52: 90%\n",
      "  bucket 54: 40%\n",
      "  bucket 56: 10%\n",
      "  bucket 55: 40%\n",
      "  bucket 53: 80%\n",
      "  bucket 52: 100%\n",
      "  Sorting block of length 56284369 for bucket 52\n",
      "  (Using difference cover)\n",
      "  bucket 54: 50%\n",
      "  Sorting block time: 00:00:59\n",
      "Returning block of 63037285 for bucket 49\n",
      "  bucket 56: 20%\n",
      "Getting block 57 of 59\n",
      "  Reserving size (67143404) for bucket 57\n",
      "  Calculating Z arrays for bucket 57\n",
      "  Entering block accumulator loop for bucket 57:\n",
      "  bucket 55: 50%\n",
      "  bucket 53: 90%\n",
      "  bucket 54: 60%\n",
      "  bucket 56: 30%\n",
      "  bucket 57: 10%\n",
      "  bucket 55: 60%\n",
      "  bucket 53: 100%\n",
      "  Sorting block of length 45383150 for bucket 53\n",
      "  (Using difference cover)\n",
      "  bucket 54: 70%\n",
      "  bucket 56: 40%\n",
      "  Sorting block time: 00:01:03\n",
      "Returning block of 65201028 for bucket 51\n",
      "  bucket 57: 20%\n",
      "Getting block 58 of 59\n",
      "  Reserving size (67143404) for bucket 58\n",
      "  Calculating Z arrays for bucket 58\n",
      "  Entering block accumulator loop for bucket 58:\n",
      "  bucket 55: 70%\n",
      "  bucket 54: 80%\n",
      "  bucket 56: 50%\n",
      "  bucket 57: 30%\n",
      "  bucket 58: 10%\n",
      "  bucket 55: 80%\n",
      "  bucket 54: 90%\n",
      "  bucket 56: 60%\n",
      "  bucket 57: 40%\n",
      "  bucket 58: 20%\n",
      "  bucket 55: 90%\n",
      "  bucket 56: 70%\n",
      "  bucket 54: 100%\n",
      "  Sorting block of length 27743820 for bucket 54\n",
      "  (Using difference cover)\n",
      "  bucket 57: 50%\n",
      "  bucket 58: 30%\n",
      "  bucket 55: 100%\n",
      "  Sorting block of length 42328305 for bucket 55\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:55\n",
      "Returning block of 56284370 for bucket 52\n",
      "  bucket 56: 80%\n",
      "Getting block 59 of 59\n",
      "  Reserving size (67143404) for bucket 59\n",
      "  Calculating Z arrays for bucket 59\n",
      "  Entering block accumulator loop for bucket 59:\n",
      "  bucket 57: 60%\n",
      "  bucket 58: 40%\n",
      "  bucket 59: 10%\n",
      "  Sorting block time: 00:00:44\n",
      "Returning block of 45383151 for bucket 53\n",
      "  bucket 56: 90%\n",
      "  bucket 59: 20%\n",
      "  bucket 57: 70%\n",
      "  bucket 58: 50%\n",
      "  bucket 59: 30%\n",
      "  Sorting block time: 00:00:25\n",
      "Returning block of 27743821 for bucket 54\n",
      "  bucket 56: 100%\n",
      "  Sorting block of length 62011881 for bucket 56\n",
      "  (Using difference cover)\n",
      "  bucket 57: 80%\n",
      "  bucket 58: 60%\n",
      "  bucket 59: 40%\n",
      "  bucket 59: 50%\n",
      "  bucket 57: 90%\n",
      "  bucket 58: 70%\n",
      "  bucket 59: 60%\n",
      "  Sorting block time: 00:00:37\n",
      "Returning block of 42328306 for bucket 55\n",
      "  bucket 59: 70%\n",
      "  bucket 57: 100%\n",
      "  Sorting block of length 48640308 for bucket 57\n",
      "  (Using difference cover)\n",
      "  bucket 58: 80%\n",
      "  bucket 59: 80%\n",
      "  bucket 58: 90%\n",
      "  bucket 59: 90%\n",
      "  bucket 59: 100%\n",
      "  Sorting block of length 59228381 for bucket 59\n",
      "  (Using difference cover)\n",
      "  bucket 58: 100%\n",
      "  Sorting block of length 67018524 for bucket 58\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:06\n",
      "Returning block of 62011882 for bucket 56\n",
      "  Sorting block time: 00:00:57\n",
      "Returning block of 48640309 for bucket 57\n",
      "  Sorting block time: 00:01:09\n",
      "Returning block of 59228382 for bucket 59\n",
      "  Sorting block time: 00:01:11\n",
      "Returning block of 67018525 for bucket 58\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 845903867\n",
      "fchr[G]: 1431613693\n",
      "fchr[T]: 2017640225\n",
      "fchr[$]: 2864785220\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 959133608 bytes to primary EBWT file: index.1.bt2\n",
      "Wrote 716196312 bytes to secondary EBWT file: index.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2864785220\n",
      "    bwtLen: 2864785221\n",
      "    sz: 716196305\n",
      "    bwtSz: 716196306\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 179049077\n",
      "    offsSz: 716196308\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14920757\n",
      "    numLines: 14920757\n",
      "    ebwtTotLen: 954928448\n",
      "    ebwtTotSz: 954928448\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:28:42\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:17\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:16\n",
      "  Time to reverse reference sequence: 00:00:02\n",
      "bmax according to bmaxDivN setting: 89524538\n",
      "Using parameters --bmax 67143404 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 67143404 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:35\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:02:47\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:42\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 86 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 86 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 86 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 13, merged 38; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 5, merged 8; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.93928e+07 (target: 67143403)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 2 of 58\n",
      "Getting block 4 of 58\n",
      "  Reserving size (67143404) for bucket 2\n",
      "  Reserving size (67143404) for bucket 4\n",
      "Getting block 6 of 58\n",
      "Getting block 7 of 58\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Reserving size (67143404) for bucket 6\n",
      "Getting block 1 of 58\n",
      "Getting block 5 of 58\n",
      "Getting block 3 of 58\n",
      "  Reserving size (67143404) for bucket 7\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  Reserving size (67143404) for bucket 1\n",
      "  Reserving size (67143404) for bucket 5\n",
      "  Reserving size (67143404) for bucket 3\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 1: 10%\n",
      "  bucket 2: 10%\n",
      "  bucket 3: 10%\n",
      "  bucket 4: 10%\n",
      "  bucket 5: 10%\n",
      "  bucket 7: 10%\n",
      "  bucket 6: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 2: 20%\n",
      "  bucket 3: 20%\n",
      "  bucket 4: 20%\n",
      "  bucket 5: 20%\n",
      "  bucket 7: 20%\n",
      "  bucket 6: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 2: 30%\n",
      "  bucket 3: 30%\n",
      "  bucket 4: 30%\n",
      "  bucket 5: 30%\n",
      "  bucket 7: 30%\n",
      "  bucket 6: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 2: 40%\n",
      "  bucket 3: 40%\n",
      "  bucket 4: 40%\n",
      "  bucket 5: 40%\n",
      "  bucket 7: 40%\n",
      "  bucket 6: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 2: 50%\n",
      "  bucket 3: 50%\n",
      "  bucket 4: 50%\n",
      "  bucket 5: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 7: 50%\n",
      "  bucket 6: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 3: 60%\n",
      "  bucket 4: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 5: 60%\n",
      "  bucket 7: 60%\n",
      "  bucket 6: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 3: 70%\n",
      "  bucket 4: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 5: 70%\n",
      "  bucket 7: 70%\n",
      "  bucket 6: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 3: 80%\n",
      "  bucket 4: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 5: 80%\n",
      "  bucket 7: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 6: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 4: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 57244230 for bucket 1\n",
      "  (Using difference cover)\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 48139330 for bucket 2\n",
      "  (Using difference cover)\n",
      "  bucket 5: 90%\n",
      "  bucket 7: 90%\n",
      "  bucket 6: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 57222742 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 59711939 for bucket 4\n",
      "  (Using difference cover)\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 65954394 for bucket 5\n",
      "  (Using difference cover)\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 24725971 for bucket 7\n",
      "  (Using difference cover)\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 52886387 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:53\n",
      "Returning block of 24725972 for bucket 7\n",
      "Getting block 8 of 58\n",
      "  Reserving size (67143404) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  Sorting block time: 00:01:24\n",
      "Returning block of 48139331 for bucket 2\n",
      "  bucket 8: 20%\n",
      "Getting block 9 of 58\n",
      "  Reserving size (67143404) for bucket 9\n",
      "  Calculating Z arrays for bucket 9\n",
      "  Entering block accumulator loop for bucket 9:\n",
      "  bucket 8: 30%\n",
      "  bucket 9: 10%\n",
      "  Sorting block time: 00:01:45\n",
      "Returning block of 57244231 for bucket 1\n",
      "  Sorting block time: 00:01:42\n",
      "Returning block of 57222743 for bucket 3\n",
      "  bucket 8: 40%\n",
      "  Sorting block time: 00:01:31\n",
      "Returning block of 52886388 for bucket 6\n",
      "Getting block 10 of 58\n",
      "  Reserving size (67143404) for bucket 10\n",
      "  Calculating Z arrays for bucket 10\n",
      "  Entering block accumulator loop for bucket 10:\n",
      "  bucket 9: 20%\n",
      "Getting block 11 of 58\n",
      "  Reserving size (67143404) for bucket 11\n",
      "  Calculating Z arrays for bucket 11\n",
      "  Entering block accumulator loop for bucket 11:\n",
      "Getting block 12 of 58\n",
      "  Reserving size (67143404) for bucket 12\n",
      "  Calculating Z arrays for bucket 12\n",
      "  Entering block accumulator loop for bucket 12:\n",
      "  Sorting block time: 00:01:42\n",
      "Returning block of 59711940 for bucket 4\n",
      "Getting block 13 of 58\n",
      "  Reserving size (67143404) for bucket 13\n",
      "  Calculating Z arrays for bucket 13\n",
      "  Entering block accumulator loop for bucket 13:\n",
      "  bucket 8: 50%\n",
      "  bucket 10: 10%\n",
      "  bucket 9: 30%\n",
      "  bucket 11: 10%\n",
      "  bucket 12: 10%\n",
      "  Sorting block time: 00:01:46\n",
      "Returning block of 65954395 for bucket 5\n",
      "  bucket 13: 10%\n",
      "Getting block 14 of 58\n",
      "  Reserving size (67143404) for bucket 14\n",
      "  Calculating Z arrays for bucket 14\n",
      "  Entering block accumulator loop for bucket 14:\n",
      "  bucket 8: 60%\n",
      "  bucket 10: 20%\n",
      "  bucket 9: 40%\n",
      "  bucket 11: 20%\n",
      "  bucket 12: 20%\n",
      "  bucket 13: 20%\n",
      "  bucket 8: 70%\n",
      "  bucket 14: 10%\n",
      "  bucket 10: 30%\n",
      "  bucket 9: 50%\n",
      "  bucket 11: 30%\n",
      "  bucket 12: 30%\n",
      "  bucket 13: 30%\n",
      "  bucket 8: 80%\n",
      "  bucket 10: 40%\n",
      "  bucket 14: 20%\n",
      "  bucket 9: 60%\n",
      "  bucket 11: 40%\n",
      "  bucket 12: 40%\n",
      "  bucket 13: 40%\n",
      "  bucket 8: 90%\n",
      "  bucket 10: 50%\n",
      "  bucket 14: 30%\n",
      "  bucket 9: 70%\n",
      "  bucket 11: 50%\n",
      "  bucket 12: 50%\n",
      "  bucket 13: 50%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 65453583 for bucket 8\n",
      "  (Using difference cover)\n",
      "  bucket 10: 60%\n",
      "  bucket 14: 40%\n",
      "  bucket 9: 80%\n",
      "  bucket 11: 60%\n",
      "  bucket 12: 60%\n",
      "  bucket 13: 60%\n",
      "  bucket 10: 70%\n",
      "  bucket 14: 50%\n",
      "  bucket 9: 90%\n",
      "  bucket 11: 70%\n",
      "  bucket 12: 70%\n",
      "  bucket 13: 70%\n",
      "  bucket 10: 80%\n",
      "  bucket 14: 60%\n",
      "  bucket 9: 100%\n",
      "  Sorting block of length 66697178 for bucket 9\n",
      "  (Using difference cover)\n",
      "  bucket 11: 80%\n",
      "  bucket 12: 80%\n",
      "  bucket 13: 80%\n",
      "  bucket 10: 90%\n",
      "  bucket 14: 70%\n",
      "  bucket 11: 90%\n",
      "  bucket 12: 90%\n",
      "  bucket 10: 100%\n",
      "  Sorting block of length 43257184 for bucket 10\n",
      "  (Using difference cover)\n",
      "  bucket 13: 90%\n",
      "  bucket 14: 80%\n",
      "  bucket 11: 100%\n",
      "  Sorting block of length 64097315 for bucket 11\n",
      "  (Using difference cover)\n",
      "  bucket 12: 100%\n",
      "  Sorting block of length 61066248 for bucket 12\n",
      "  (Using difference cover)\n",
      "  bucket 13: 100%\n",
      "  Sorting block of length 37042782 for bucket 13\n",
      "  (Using difference cover)\n",
      "  bucket 14: 90%\n",
      "  bucket 14: 100%\n",
      "  Sorting block of length 56613405 for bucket 14\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:38\n",
      "Returning block of 65453584 for bucket 8\n",
      "  Sorting block time: 00:00:57\n",
      "Returning block of 37042783 for bucket 13\n",
      "  Sorting block time: 00:01:07\n",
      "Returning block of 43257185 for bucket 10\n",
      "Getting block 15 of 58\n",
      "  Reserving size (67143404) for bucket 15\n",
      "  Calculating Z arrays for bucket 15\n",
      "  Entering block accumulator loop for bucket 15:\n",
      "Getting block 16 of 58\n",
      "  Reserving size (67143404) for bucket 16\n",
      "  Calculating Z arrays for bucket 16\n",
      "  Entering block accumulator loop for bucket 16:\n",
      "Getting block 17 of 58\n",
      "  Reserving size (67143404) for bucket 17\n",
      "  Calculating Z arrays for bucket 17\n",
      "  Entering block accumulator loop for bucket 17:\n",
      "  bucket 15: 10%\n",
      "  bucket 16: 10%\n",
      "  bucket 17: 10%\n",
      "  Sorting block time: 00:01:36\n",
      "Returning block of 66697179 for bucket 9\n",
      "  bucket 15: 20%\n",
      "  bucket 17: 20%\n",
      "  bucket 16: 20%\n",
      "Getting block 18 of 58\n",
      "  Reserving size (67143404) for bucket 18\n",
      "  Calculating Z arrays for bucket 18\n",
      "  Entering block accumulator loop for bucket 18:\n",
      "  bucket 15: 30%\n",
      "  bucket 17: 30%\n",
      "  bucket 16: 30%\n",
      "  bucket 18: 10%\n",
      "  Sorting block time: 00:01:28\n",
      "Returning block of 61066249 for bucket 12\n",
      "Getting block 19 of 58\n",
      "  Reserving size (67143404) for bucket 19\n",
      "  Calculating Z arrays for bucket 19\n",
      "  Entering block accumulator loop for bucket 19:\n",
      "  Sorting block time: 00:01:32\n",
      "Returning block of 64097316 for bucket 11\n",
      "  Sorting block time: 00:01:18\n",
      "Returning block of 56613406 for bucket 14\n",
      "  bucket 15: 40%\n",
      "  bucket 17: 40%\n",
      "Getting block 20 of 58\n",
      "  Reserving size (67143404) for bucket 20\n",
      "  Calculating Z arrays for bucket 20\n",
      "  Entering block accumulator loop for bucket 20:\n",
      "Getting block 21 of 58\n",
      "  Reserving size (67143404) for bucket 21\n",
      "  Calculating Z arrays for bucket 21\n",
      "  Entering block accumulator loop for bucket 21:\n",
      "  bucket 16: 40%\n",
      "  bucket 18: 20%\n",
      "  bucket 19: 10%\n",
      "  bucket 15: 50%\n",
      "  bucket 17: 50%\n",
      "  bucket 20: 10%\n",
      "  bucket 21: 10%\n",
      "  bucket 18: 30%\n",
      "  bucket 16: 50%\n",
      "  bucket 19: 20%\n",
      "  bucket 15: 60%\n",
      "  bucket 17: 60%\n",
      "  bucket 20: 20%\n",
      "  bucket 21: 20%\n",
      "  bucket 18: 40%\n",
      "  bucket 16: 60%\n",
      "  bucket 19: 30%\n",
      "  bucket 15: 70%\n",
      "  bucket 17: 70%\n",
      "  bucket 20: 30%\n",
      "  bucket 21: 30%\n",
      "  bucket 18: 50%\n",
      "  bucket 16: 70%\n",
      "  bucket 19: 40%\n",
      "  bucket 15: 80%\n",
      "  bucket 17: 80%\n",
      "  bucket 20: 40%\n",
      "  bucket 21: 40%\n",
      "  bucket 18: 60%\n",
      "  bucket 16: 80%\n",
      "  bucket 19: 50%\n",
      "  bucket 15: 90%\n",
      "  bucket 17: 90%\n",
      "  bucket 20: 50%\n",
      "  bucket 21: 50%\n",
      "  bucket 18: 70%\n",
      "  bucket 16: 90%\n",
      "  bucket 19: 60%\n",
      "  bucket 17: 100%\n",
      "  Sorting block of length 10792698 for bucket 17\n",
      "  (Using difference cover)\n",
      "  bucket 15: 100%\n",
      "  Sorting block of length 60489522 for bucket 15\n",
      "  (Using difference cover)\n",
      "  bucket 20: 60%\n",
      "  bucket 18: 80%\n",
      "  bucket 21: 60%\n",
      "  bucket 16: 100%\n",
      "  Sorting block of length 57610739 for bucket 16\n",
      "  (Using difference cover)\n",
      "  bucket 19: 70%\n",
      "  bucket 20: 70%\n",
      "  bucket 18: 90%\n",
      "  bucket 21: 70%\n",
      "  Sorting block time: 00:00:16\n",
      "Returning block of 10792699 for bucket 17\n",
      "Getting block 22 of 58\n",
      "  Reserving size (67143404) for bucket 22\n",
      "  Calculating Z arrays for bucket 22\n",
      "  Entering block accumulator loop for bucket 22:\n",
      "  bucket 19: 80%\n",
      "  bucket 20: 80%\n",
      "  bucket 18: 100%\n",
      "  Sorting block of length 62230214 for bucket 18\n",
      "  (Using difference cover)\n",
      "  bucket 21: 80%\n",
      "  bucket 22: 10%\n",
      "  bucket 19: 90%\n",
      "  bucket 20: 90%\n",
      "  bucket 21: 90%\n",
      "  bucket 19: 100%\n",
      "  Sorting block of length 43304729 for bucket 19\n",
      "  (Using difference cover)\n",
      "  bucket 22: 20%\n",
      "  bucket 20: 100%\n",
      "  Sorting block of length 64890512 for bucket 20\n",
      "  (Using difference cover)\n",
      "  bucket 21: 100%\n",
      "  Sorting block of length 66586703 for bucket 21\n",
      "  (Using difference cover)\n",
      "  bucket 22: 30%\n",
      "  bucket 22: 40%\n",
      "  bucket 22: 50%\n",
      "  bucket 22: 60%\n",
      "  Sorting block time: 00:01:27\n",
      "Returning block of 60489523 for bucket 15\n",
      "  bucket 22: 70%\n",
      "  Sorting block time: 00:01:21\n",
      "Returning block of 57610740 for bucket 16\n",
      "Getting block 23 of 58\n",
      "  Reserving size (67143404) for bucket 23\n",
      "  Calculating Z arrays for bucket 23\n",
      "  Entering block accumulator loop for bucket 23:\n",
      "Getting block 24 of 58\n",
      "  Reserving size (67143404) for bucket 24\n",
      "  Calculating Z arrays for bucket 24\n",
      "  Entering block accumulator loop for bucket 24:\n",
      "  bucket 22: 80%\n",
      "  Sorting block time: 00:00:59\n",
      "Returning block of 43304730 for bucket 19\n",
      "  bucket 23: 10%\n",
      "Getting block 25 of 58\n",
      "  Reserving size (67143404) for bucket 25\n",
      "  Calculating Z arrays for bucket 25\n",
      "  Entering block accumulator loop for bucket 25:\n",
      "  bucket 24: 10%\n",
      "  bucket 22: 90%\n",
      "  bucket 23: 20%\n",
      "  bucket 25: 10%\n",
      "  bucket 24: 20%\n",
      "  Sorting block time: 00:01:28\n",
      "Returning block of 62230215 for bucket 18\n",
      "  bucket 22: 100%\n",
      "  Sorting block of length 16719129 for bucket 22\n",
      "  (Using difference cover)\n",
      "Getting block 26 of 58\n",
      "  Reserving size (67143404) for bucket 26\n",
      "  Calculating Z arrays for bucket 26\n",
      "  Entering block accumulator loop for bucket 26:\n",
      "  bucket 23: 30%\n",
      "  bucket 25: 20%\n",
      "  bucket 24: 30%\n",
      "  bucket 26: 10%\n",
      "  bucket 23: 40%\n",
      "  bucket 25: 30%\n",
      "  Sorting block time: 00:01:24\n",
      "Returning block of 64890513 for bucket 20\n",
      "  bucket 24: 40%\n",
      "Getting block 27 of 58\n",
      "  Reserving size (67143404) for bucket 27\n",
      "  Calculating Z arrays for bucket 27\n",
      "  Entering block accumulator loop for bucket 27:\n",
      "  Sorting block time: 00:00:17\n",
      "Returning block of 16719130 for bucket 22\n",
      "  bucket 26: 20%\n",
      "Getting block 28 of 58\n",
      "  Reserving size (67143404) for bucket 28\n",
      "  Calculating Z arrays for bucket 28\n",
      "  Entering block accumulator loop for bucket 28:\n",
      "  bucket 23: 50%\n",
      "  Sorting block time: 00:01:28\n",
      "Returning block of 66586704 for bucket 21\n",
      "  bucket 25: 40%\n",
      "  bucket 24: 50%\n",
      "Getting block 29 of 58\n",
      "  Reserving size (67143404) for bucket 29\n",
      "  Calculating Z arrays for bucket 29\n",
      "  Entering block accumulator loop for bucket 29:\n",
      "  bucket 27: 10%\n",
      "  bucket 26: 30%\n",
      "  bucket 28: 10%\n",
      "  bucket 23: 60%\n",
      "  bucket 25: 50%\n",
      "  bucket 24: 60%\n",
      "  bucket 29: 10%\n",
      "  bucket 27: 20%\n",
      "  bucket 26: 40%\n",
      "  bucket 28: 20%\n",
      "  bucket 23: 70%\n",
      "  bucket 25: 60%\n",
      "  bucket 24: 70%\n",
      "  bucket 29: 20%\n",
      "  bucket 27: 30%\n",
      "  bucket 26: 50%\n",
      "  bucket 28: 30%\n",
      "  bucket 23: 80%\n",
      "  bucket 24: 80%\n",
      "  bucket 25: 70%\n",
      "  bucket 26: 60%\n",
      "  bucket 27: 40%\n",
      "  bucket 29: 30%\n",
      "  bucket 28: 40%\n",
      "  bucket 23: 90%\n",
      "  bucket 24: 90%\n",
      "  bucket 25: 80%\n",
      "  bucket 26: 70%\n",
      "  bucket 27: 50%\n",
      "  bucket 29: 40%\n",
      "  bucket 28: 50%\n",
      "  bucket 23: 100%\n",
      "  Sorting block of length 64617932 for bucket 23\n",
      "  (Using difference cover)\n",
      "  bucket 24: 100%\n",
      "  Sorting block of length 34830437 for bucket 24\n",
      "  (Using difference cover)\n",
      "  bucket 25: 90%\n",
      "  bucket 26: 80%\n",
      "  bucket 27: 60%\n",
      "  bucket 28: 60%\n",
      "  bucket 29: 50%\n",
      "  bucket 25: 100%\n",
      "  Sorting block of length 37658512 for bucket 25\n",
      "  (Using difference cover)\n",
      "  bucket 26: 90%\n",
      "  bucket 27: 70%\n",
      "  bucket 28: 70%\n",
      "  bucket 29: 60%\n",
      "  bucket 26: 100%\n",
      "  Sorting block of length 52236532 for bucket 26\n",
      "  (Using difference cover)\n",
      "  bucket 27: 80%\n",
      "  bucket 28: 80%\n",
      "  bucket 29: 70%\n",
      "  bucket 27: 90%\n",
      "  bucket 28: 90%\n",
      "  bucket 29: 80%\n",
      "  Sorting block time: 00:00:37\n",
      "Returning block of 34830438 for bucket 24\n",
      "Getting block 30 of 58\n",
      "  Reserving size (67143404) for bucket 30\n",
      "  Calculating Z arrays for bucket 30\n",
      "  Entering block accumulator loop for bucket 30:\n",
      "  bucket 27: 100%\n",
      "  Sorting block of length 65566231 for bucket 27\n",
      "  (Using difference cover)\n",
      "  bucket 28: 100%\n",
      "  Sorting block of length 29550056 for bucket 28\n",
      "  (Using difference cover)\n",
      "  bucket 29: 90%\n",
      "  bucket 30: 10%\n",
      "  Sorting block time: 00:00:38\n",
      "Returning block of 37658513 for bucket 25\n",
      "Getting block 31 of 58\n",
      "  Reserving size (67143404) for bucket 31\n",
      "  Calculating Z arrays for bucket 31\n",
      "  Entering block accumulator loop for bucket 31:\n",
      "  bucket 29: 100%\n",
      "  Sorting block of length 57248503 for bucket 29\n",
      "  (Using difference cover)\n",
      "  bucket 30: 20%\n",
      "  bucket 31: 10%\n",
      "  bucket 30: 30%\n",
      "  bucket 31: 20%\n",
      "  Sorting block time: 00:01:15\n",
      "Returning block of 64617933 for bucket 23\n",
      "  Sorting block time: 00:00:33\n",
      "Returning block of 29550057 for bucket 28\n",
      "  bucket 30: 40%\n",
      "Getting block 32 of 58\n",
      "  Reserving size (67143404) for bucket 32\n",
      "  Calculating Z arrays for bucket 32\n",
      "  Entering block accumulator loop for bucket 32:\n",
      "Getting block 33 of 58\n",
      "  Reserving size (67143404) for bucket 33\n",
      "  Calculating Z arrays for bucket 33\n",
      "  Entering block accumulator loop for bucket 33:\n",
      "  bucket 31: 30%\n",
      "  Sorting block time: 00:00:57\n",
      "Returning block of 52236533 for bucket 26\n",
      "  bucket 30: 50%\n",
      "Getting block 34 of 58\n",
      "  Reserving size (67143404) for bucket 34\n",
      "  Calculating Z arrays for bucket 34\n",
      "  Entering block accumulator loop for bucket 34:\n",
      "  bucket 32: 10%\n",
      "  bucket 33: 10%\n",
      "  bucket 31: 40%\n",
      "  bucket 30: 60%\n",
      "  bucket 34: 10%\n",
      "  bucket 32: 20%\n",
      "  bucket 33: 20%\n",
      "  bucket 31: 50%\n",
      "  bucket 30: 70%\n",
      "  bucket 34: 20%\n",
      "  bucket 32: 30%\n",
      "  bucket 33: 30%\n",
      "  bucket 31: 60%\n",
      "  bucket 30: 80%\n",
      "  bucket 34: 30%\n",
      "  bucket 32: 40%\n",
      "  bucket 33: 40%\n",
      "  Sorting block time: 00:01:14\n",
      "Returning block of 65566232 for bucket 27\n",
      "  Sorting block time: 00:01:02\n",
      "Returning block of 57248504 for bucket 29\n",
      "  bucket 31: 70%\n",
      "Getting block 35 of 58\n",
      "  Reserving size (67143404) for bucket 35\n",
      "  Calculating Z arrays for bucket 35\n",
      "  Entering block accumulator loop for bucket 35:\n",
      "Getting block 36 of 58\n",
      "  Reserving size (67143404) for bucket 36\n",
      "  Calculating Z arrays for bucket 36\n",
      "  Entering block accumulator loop for bucket 36:\n",
      "  bucket 34: 40%\n",
      "  bucket 30: 90%\n",
      "  bucket 32: 50%\n",
      "  bucket 33: 50%\n",
      "  bucket 31: 80%\n",
      "  bucket 35: 10%\n",
      "  bucket 36: 10%\n",
      "  bucket 34: 50%\n",
      "  bucket 30: 100%\n",
      "  Sorting block of length 58531017 for bucket 30\n",
      "  (Using difference cover)\n",
      "  bucket 32: 60%\n",
      "  bucket 33: 60%\n",
      "  bucket 31: 90%\n",
      "  bucket 35: 20%\n",
      "  bucket 34: 60%\n",
      "  bucket 36: 20%\n",
      "  bucket 32: 70%\n",
      "  bucket 33: 70%\n",
      "  bucket 31: 100%\n",
      "  Sorting block of length 41935353 for bucket 31\n",
      "  (Using difference cover)\n",
      "  bucket 35: 30%\n",
      "  bucket 34: 70%\n",
      "  bucket 36: 30%\n",
      "  bucket 33: 80%\n",
      "  bucket 32: 80%\n",
      "  bucket 35: 40%\n",
      "  bucket 34: 80%\n",
      "  bucket 36: 40%\n",
      "  bucket 33: 90%\n",
      "  bucket 32: 90%\n",
      "  bucket 35: 50%\n",
      "  bucket 34: 90%\n",
      "  bucket 36: 50%\n",
      "  bucket 32: 100%\n",
      "  Sorting block of length 66578378 for bucket 32\n",
      "  (Using difference cover)\n",
      "  bucket 33: 100%\n",
      "  Sorting block of length 53465738 for bucket 33\n",
      "  (Using difference cover)\n",
      "  bucket 35: 60%\n",
      "  bucket 34: 100%\n",
      "  Sorting block of length 21879593 for bucket 34\n",
      "  (Using difference cover)\n",
      "  bucket 36: 60%\n",
      "  bucket 35: 70%\n",
      "  bucket 36: 70%\n",
      "  bucket 35: 80%\n",
      "  Sorting block time: 00:00:46\n",
      "Returning block of 41935354 for bucket 31\n",
      "  Sorting block time: 00:01:02\n",
      "Returning block of 58531018 for bucket 30\n",
      "Getting block 37 of 58\n",
      "  Reserving size (67143404) for bucket 37\n",
      "  Calculating Z arrays for bucket 37\n",
      "  Entering block accumulator loop for bucket 37:\n",
      "  bucket 36: 80%\n",
      "Getting block 38 of 58\n",
      "  Reserving size (67143404) for bucket 38\n",
      "  Calculating Z arrays for bucket 38\n",
      "  Entering block accumulator loop for bucket 38:\n",
      "  Sorting block time: 00:00:22\n",
      "Returning block of 21879594 for bucket 34\n",
      "Getting block 39 of 58\n",
      "  Reserving size (67143404) for bucket 39\n",
      "  Calculating Z arrays for bucket 39\n",
      "  Entering block accumulator loop for bucket 39:\n",
      "  bucket 35: 90%\n",
      "  bucket 37: 10%\n",
      "  bucket 36: 90%\n",
      "  bucket 38: 10%\n",
      "  bucket 39: 10%\n",
      "  bucket 35: 100%\n",
      "  Sorting block of length 47396753 for bucket 35\n",
      "  (Using difference cover)\n",
      "  bucket 37: 20%\n",
      "  bucket 36: 100%\n",
      "  Sorting block of length 55264338 for bucket 36\n",
      "  (Using difference cover)\n",
      "  bucket 38: 20%\n",
      "  bucket 39: 20%\n",
      "  bucket 37: 30%\n",
      "  bucket 38: 30%\n",
      "  bucket 39: 30%\n",
      "  Sorting block time: 00:00:59\n",
      "Returning block of 53465739 for bucket 33\n",
      "Getting block 40 of 58\n",
      "  Reserving size (67143404) for bucket 40\n",
      "  Calculating Z arrays for bucket 40\n",
      "  Entering block accumulator loop for bucket 40:\n",
      "  bucket 37: 40%\n",
      "  bucket 38: 40%\n",
      "  bucket 39: 40%\n",
      "  bucket 37: 50%\n",
      "  bucket 40: 10%\n",
      "  bucket 38: 50%\n",
      "  Sorting block time: 00:01:15\n",
      "Returning block of 66578379 for bucket 32\n",
      "  bucket 39: 50%\n",
      "Getting block 41 of 58\n",
      "  Reserving size (67143404) for bucket 41\n",
      "  Calculating Z arrays for bucket 41\n",
      "  Entering block accumulator loop for bucket 41:\n",
      "  bucket 37: 60%\n",
      "  bucket 38: 60%\n",
      "  bucket 40: 20%\n",
      "  bucket 39: 60%\n",
      "  bucket 41: 10%\n",
      "  bucket 37: 70%\n",
      "  bucket 38: 70%\n",
      "  Sorting block time: 00:00:57\n",
      "Returning block of 47396754 for bucket 35\n",
      "  bucket 40: 30%\n",
      "  bucket 39: 70%\n",
      "Getting block 42 of 58\n",
      "  Reserving size (67143404) for bucket 42\n",
      "  Calculating Z arrays for bucket 42\n",
      "  Entering block accumulator loop for bucket 42:\n",
      "  bucket 41: 20%\n",
      "  bucket 37: 80%\n",
      "  bucket 38: 80%\n",
      "  bucket 40: 40%\n",
      "  bucket 39: 80%\n",
      "  bucket 42: 10%\n",
      "  bucket 41: 30%\n",
      "  Sorting block time: 00:01:11\n",
      "Returning block of 55264339 for bucket 36\n",
      "  bucket 37: 90%\n",
      "  bucket 38: 90%\n",
      "Getting block 43 of 58\n",
      "  Reserving size (67143404) for bucket 43\n",
      "  Calculating Z arrays for bucket 43\n",
      "  Entering block accumulator loop for bucket 43:\n",
      "  bucket 42: 20%\n",
      "  bucket 39: 90%\n",
      "  bucket 40: 50%\n",
      "  bucket 41: 40%\n",
      "  bucket 37: 100%\n",
      "  Sorting block of length 43717188 for bucket 37\n",
      "  (Using difference cover)\n",
      "  bucket 38: 100%\n",
      "  Sorting block of length 53562217 for bucket 38\n",
      "  (Using difference cover)\n",
      "  bucket 43: 10%\n",
      "  bucket 42: 30%\n",
      "  bucket 39: 100%\n",
      "  Sorting block of length 32754379 for bucket 39\n",
      "  (Using difference cover)\n",
      "  bucket 41: 50%\n",
      "  bucket 40: 60%\n",
      "  bucket 43: 20%\n",
      "  bucket 42: 40%\n",
      "  bucket 41: 60%\n",
      "  bucket 40: 70%\n",
      "  bucket 43: 30%\n",
      "  bucket 42: 50%\n",
      "  bucket 41: 70%\n",
      "  bucket 40: 80%\n",
      "  bucket 43: 40%\n",
      "  bucket 42: 60%\n",
      "  bucket 41: 80%\n",
      "  bucket 40: 90%\n",
      "  bucket 42: 70%\n",
      "  bucket 43: 50%\n",
      "  Sorting block time: 00:00:39\n",
      "Returning block of 32754380 for bucket 39\n",
      "  bucket 41: 90%\n",
      "Getting block 44 of 58\n",
      "  Reserving size (67143404) for bucket 44\n",
      "  Calculating Z arrays for bucket 44\n",
      "  Entering block accumulator loop for bucket 44:\n",
      "  bucket 40: 100%\n",
      "  Sorting block of length 64595703 for bucket 40\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:50\n",
      "Returning block of 43717189 for bucket 37\n",
      "  bucket 42: 80%\n",
      "  bucket 43: 60%\n",
      "Getting block 45 of 58\n",
      "  Reserving size (67143404) for bucket 45\n",
      "  Calculating Z arrays for bucket 45\n",
      "  Entering block accumulator loop for bucket 45:\n",
      "  bucket 41: 100%\n",
      "  Sorting block of length 60590139 for bucket 41\n",
      "  (Using difference cover)\n",
      "  bucket 44: 10%\n",
      "  bucket 42: 90%\n",
      "  bucket 43: 70%\n",
      "  bucket 45: 10%\n",
      "  bucket 44: 20%\n",
      "  Sorting block time: 00:01:06\n",
      "Returning block of 53562218 for bucket 38\n",
      "Getting block 46 of 58\n",
      "  Reserving size (67143404) for bucket 46\n",
      "  Calculating Z arrays for bucket 46\n",
      "  Entering block accumulator loop for bucket 46:\n",
      "  bucket 42: 100%\n",
      "  Sorting block of length 19275775 for bucket 42\n",
      "  (Using difference cover)\n",
      "  bucket 43: 80%\n",
      "  bucket 45: 20%\n",
      "  bucket 44: 30%\n",
      "  bucket 46: 10%\n",
      "  bucket 43: 90%\n",
      "  bucket 45: 30%\n",
      "  bucket 44: 40%\n",
      "  bucket 46: 20%\n",
      "  Sorting block time: 00:00:22\n",
      "Returning block of 19275776 for bucket 42\n",
      "  bucket 43: 100%\n",
      "  Sorting block of length 60996395 for bucket 43\n",
      "  (Using difference cover)\n",
      "Getting block 47 of 58\n",
      "  Reserving size (67143404) for bucket 47\n",
      "  Calculating Z arrays for bucket 47\n",
      "  Entering block accumulator loop for bucket 47:\n",
      "  bucket 45: 40%\n",
      "  bucket 44: 50%\n",
      "  bucket 46: 30%\n",
      "  bucket 47: 10%\n",
      "  bucket 45: 50%\n",
      "  bucket 44: 60%\n",
      "  bucket 46: 40%\n",
      "  bucket 45: 60%\n",
      "  bucket 47: 20%\n",
      "  bucket 44: 70%\n",
      "  bucket 46: 50%\n",
      "  Sorting block time: 00:01:14\n",
      "Returning block of 60590140 for bucket 41\n",
      "  bucket 45: 70%\n",
      "  bucket 47: 30%\n",
      "  bucket 44: 80%\n",
      "Getting block 48 of 58\n",
      "  Reserving size (67143404) for bucket 48\n",
      "  Calculating Z arrays for bucket 48\n",
      "  Entering block accumulator loop for bucket 48:\n",
      "  Sorting block time: 00:01:24\n",
      "Returning block of 64595704 for bucket 40\n",
      "Getting block 49 of 58\n",
      "  Reserving size (67143404) for bucket 49\n",
      "  Calculating Z arrays for bucket 49\n",
      "  Entering block accumulator loop for bucket 49:\n",
      "  bucket 46: 60%\n",
      "  bucket 45: 80%\n",
      "  bucket 47: 40%\n",
      "  bucket 44: 90%\n",
      "  bucket 48: 10%\n",
      "  bucket 49: 10%\n",
      "  bucket 46: 70%\n",
      "  bucket 45: 90%\n",
      "  bucket 47: 50%\n",
      "  bucket 44: 100%\n",
      "  Sorting block of length 64857419 for bucket 44\n",
      "  (Using difference cover)\n",
      "  bucket 48: 20%\n",
      "  bucket 49: 20%\n",
      "  bucket 46: 80%\n",
      "  bucket 45: 100%\n",
      "  Sorting block of length 47604684 for bucket 45\n",
      "  (Using difference cover)\n",
      "  bucket 47: 60%\n",
      "  bucket 48: 30%\n",
      "  Sorting block time: 00:01:10\n",
      "Returning block of 60996396 for bucket 43\n",
      "  bucket 49: 30%\n",
      "  bucket 46: 90%\n",
      "Getting block 50 of 58\n",
      "  Reserving size (67143404) for bucket 50\n",
      "  Calculating Z arrays for bucket 50\n",
      "  Entering block accumulator loop for bucket 50:\n",
      "  bucket 47: 70%\n",
      "  bucket 48: 40%\n",
      "  bucket 46: 100%\n",
      "  Sorting block of length 25288166 for bucket 46\n",
      "  (Using difference cover)\n",
      "  bucket 49: 40%\n",
      "  bucket 50: 10%\n",
      "  bucket 47: 80%\n",
      "  bucket 48: 50%\n",
      "  bucket 49: 50%\n",
      "  bucket 50: 20%\n",
      "  bucket 47: 90%\n",
      "  bucket 48: 60%\n",
      "  bucket 49: 60%\n",
      "  bucket 50: 30%\n",
      "  bucket 48: 70%\n",
      "  bucket 47: 100%\n",
      "  Sorting block of length 65382233 for bucket 47\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:26\n",
      "Returning block of 25288167 for bucket 46\n",
      "Getting block 51 of 58\n",
      "  Reserving size (67143404) for bucket 51\n",
      "  Calculating Z arrays for bucket 51\n",
      "  Entering block accumulator loop for bucket 51:\n",
      "  bucket 49: 70%\n",
      "  bucket 50: 40%\n",
      "  Sorting block time: 00:00:49\n",
      "Returning block of 47604685 for bucket 45\n",
      "  bucket 48: 80%\n",
      "Getting block 52 of 58\n",
      "  Reserving size (67143404) for bucket 52\n",
      "  Calculating Z arrays for bucket 52\n",
      "  Entering block accumulator loop for bucket 52:\n",
      "  bucket 51: 10%\n",
      "  bucket 49: 80%\n",
      "  bucket 50: 50%\n",
      "  Sorting block time: 00:01:04\n",
      "Returning block of 64857420 for bucket 44\n",
      "  bucket 52: 10%\n",
      "Getting block 53 of 58\n",
      "  Reserving size (67143404) for bucket 53\n",
      "  Calculating Z arrays for bucket 53\n",
      "  Entering block accumulator loop for bucket 53:\n",
      "  bucket 48: 90%\n",
      "  bucket 51: 20%\n",
      "  bucket 49: 90%\n",
      "  bucket 50: 60%\n",
      "  bucket 52: 20%\n",
      "  bucket 53: 10%\n",
      "  bucket 48: 100%\n",
      "  Sorting block of length 65206773 for bucket 48\n",
      "  (Using difference cover)\n",
      "  bucket 51: 30%\n",
      "  bucket 49: 100%\n",
      "  Sorting block of length 44876051 for bucket 49\n",
      "  (Using difference cover)\n",
      "  bucket 50: 70%\n",
      "  bucket 53: 20%\n",
      "  bucket 52: 30%\n",
      "  bucket 51: 40%\n",
      "  bucket 50: 80%\n",
      "  bucket 52: 40%\n",
      "  bucket 53: 30%\n",
      "  bucket 51: 50%\n",
      "  bucket 50: 90%\n",
      "  bucket 52: 50%\n",
      "  bucket 53: 40%\n",
      "  bucket 51: 60%\n",
      "  bucket 50: 100%\n",
      "  Sorting block of length 64403433 for bucket 50\n",
      "  (Using difference cover)\n",
      "  bucket 52: 60%\n",
      "  bucket 53: 50%\n",
      "  bucket 51: 70%\n",
      "  bucket 52: 70%\n",
      "  bucket 53: 60%\n",
      "  bucket 51: 80%\n",
      "  bucket 52: 80%\n",
      "  bucket 53: 70%\n",
      "  bucket 51: 90%\n",
      "  bucket 52: 90%\n",
      "  bucket 53: 80%\n",
      "  bucket 51: 100%\n",
      "  Sorting block of length 62330760 for bucket 51\n",
      "  (Using difference cover)\n",
      "  bucket 52: 100%\n",
      "  Sorting block of length 29292166 for bucket 52\n",
      "  (Using difference cover)\n",
      "  bucket 53: 90%\n",
      "  bucket 53: 100%\n",
      "  Sorting block of length 50074017 for bucket 53\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:04:24\n",
      "Returning block of 65382234 for bucket 47\n",
      "Getting block 54 of 58\n",
      "  Reserving size (67143404) for bucket 54\n",
      "  Calculating Z arrays for bucket 54\n",
      "  Entering block accumulator loop for bucket 54:\n",
      "  bucket 54: 10%\n",
      "  Sorting block time: 00:04:18\n",
      "Returning block of 44876052 for bucket 49\n",
      "Getting block 55 of 58\n",
      "  Reserving size (67143404) for bucket 55\n",
      "  Calculating Z arrays for bucket 55\n",
      "  Entering block accumulator loop for bucket 55:\n",
      "  bucket 54: 20%\n",
      "  bucket 55: 10%\n",
      "  Sorting block time: 00:01:20\n",
      "Returning block of 29292167 for bucket 52\n",
      "  bucket 54: 30%\n",
      "Getting block 56 of 58\n",
      "  Reserving size (67143404) for bucket 56\n",
      "  Calculating Z arrays for bucket 56\n",
      "  Entering block accumulator loop for bucket 56:\n",
      "  Sorting block time: 00:04:47\n",
      "Returning block of 65206774 for bucket 48\n",
      "  bucket 55: 20%\n",
      "Getting block 57 of 58\n",
      "  Reserving size (67143404) for bucket 57\n",
      "  Calculating Z arrays for bucket 57\n",
      "  Entering block accumulator loop for bucket 57:\n",
      "  bucket 54: 40%\n",
      "  bucket 56: 10%\n",
      "  bucket 55: 30%\n",
      "  bucket 57: 10%\n",
      "  bucket 54: 50%\n",
      "  bucket 56: 20%\n",
      "  bucket 55: 40%\n",
      "  bucket 57: 20%\n",
      "  Sorting block time: 00:04:15\n",
      "Returning block of 64403434 for bucket 50\n",
      "  bucket 54: 60%\n",
      "  bucket 56: 30%\n",
      "Getting block 58 of 58\n",
      "  Reserving size (67143404) for bucket 58\n",
      "  Calculating Z arrays for bucket 58\n",
      "  Entering block accumulator loop for bucket 58:\n",
      "  bucket 55: 50%\n",
      "  bucket 57: 30%\n",
      "  bucket 58: 10%\n",
      "  bucket 54: 70%\n",
      "  Sorting block time: 00:02:18\n",
      "Returning block of 50074018 for bucket 53\n",
      "  bucket 56: 40%\n",
      "  bucket 58: 20%\n",
      "  bucket 55: 60%\n",
      "  bucket 57: 40%\n",
      "  Sorting block time: 00:02:57\n",
      "Returning block of 62330761 for bucket 51\n",
      "  bucket 54: 80%\n",
      "  bucket 56: 50%\n",
      "  bucket 58: 30%\n",
      "  bucket 55: 70%\n",
      "  bucket 57: 50%\n",
      "  bucket 58: 40%\n",
      "  bucket 54: 90%\n",
      "  bucket 56: 60%\n",
      "  bucket 58: 50%\n",
      "  bucket 57: 60%\n",
      "  bucket 55: 80%\n",
      "  bucket 58: 60%\n",
      "  bucket 54: 100%\n",
      "  Sorting block of length 17593708 for bucket 54\n",
      "  (Using difference cover)\n",
      "  bucket 56: 70%\n",
      "  bucket 58: 70%\n",
      "  bucket 57: 70%\n",
      "  bucket 55: 90%\n",
      "  bucket 56: 80%\n",
      "  bucket 58: 80%\n",
      "  bucket 57: 80%\n",
      "  bucket 58: 90%\n",
      "  bucket 55: 100%\n",
      "  Sorting block of length 58230575 for bucket 55\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:19\n",
      "Returning block of 17593709 for bucket 54\n",
      "  bucket 56: 90%\n",
      "  bucket 58: 100%\n",
      "  Sorting block of length 31578271 for bucket 58\n",
      "  (Using difference cover)\n",
      "  bucket 57: 90%\n",
      "  bucket 56: 100%\n",
      "  Sorting block of length 37411433 for bucket 56\n",
      "  (Using difference cover)\n",
      "  bucket 57: 100%\n",
      "  Sorting block of length 35667371 for bucket 57\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:34\n",
      "Returning block of 31578272 for bucket 58\n",
      "  Sorting block time: 00:00:40\n",
      "Returning block of 37411434 for bucket 56\n",
      "  Sorting block time: 00:00:37\n",
      "Returning block of 35667372 for bucket 57\n",
      "  Sorting block time: 00:01:00\n",
      "Returning block of 58230576 for bucket 55\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 845903867\n",
      "fchr[G]: 1431613693\n",
      "fchr[T]: 2017640225\n",
      "fchr[$]: 2864785220\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 959133608 bytes to primary EBWT file: index.rev.1.bt2\n",
      "Wrote 716196312 bytes to secondary EBWT file: index.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2864785220\n",
      "    bwtLen: 2864785221\n",
      "    sz: 716196305\n",
      "    bwtSz: 716196306\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 179049077\n",
      "    offsSz: 716196308\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 14920757\n",
      "    numLines: 14920757\n",
      "    ebwtTotLen: 954928448\n",
      "    ebwtTotSz: 954928448\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:34:37\n"
     ]
    }
   ],
   "source": [
    "! bowtie2-build $mypath/reference.fa index --threads 8\n",
    "#maybe try to output bampe files next time (and with higher sensibility) for macs2 to read the paired end information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### we then produce the alignments\n",
    "\n",
    "here we use the paired end alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time loading reference: 00:00:00\n",
      "Time loading forward index: 00:00:01\n",
      "Time loading mirror index: 00:00:03\n",
      "Multiseed full-index search: 01:42:07\n",
      "29050599 reads; of these:\n",
      "  29050599 (100.00%) were paired; of these:\n",
      "    8460207 (29.12%) aligned concordantly 0 times\n",
      "    18598191 (64.02%) aligned concordantly exactly 1 time\n",
      "    1992201 (6.86%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8460207 pairs aligned concordantly 0 times; of these:\n",
      "      905301 (10.70%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7554906 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15109812 mates make up the pairs; of these:\n",
      "        14345991 (94.94%) aligned 0 times\n",
      "        452282 (2.99%) aligned exactly 1 time\n",
      "        311539 (2.06%) aligned >1 times\n",
      "75.31% overall alignment rate\n",
      "Time searching: 01:42:12\n",
      "Overall time: 01:42:12\n"
     ]
    }
   ],
   "source": [
    "! bowtie2 -x index -1 $mypath/54-mp457-PDX34-H3K27ac-r2_S7_R1.fastq.gz -2 $mypath/54-mp457-PDX34-H3K27ac-r2_S7_R2.fastq.gz -S $mypath/mapped.sam --threads 8 -t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### we can then use macs2 for the peak calling "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: MACS2: command not found\r\n"
     ]
    }
   ],
   "source": [
    "! MACS2 callpeak -f BAMPE -t $mypath/mapped.sam --outdir \"peak_called_bampe\"\n",
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some visualization..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \r\n",
      "          1 \r\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"./peak_called/NA_model.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x10694ae90>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "! Rscript peak_called/NA_model.r\n",
    "from IPython.display import IFrame\n",
    "IFrame(\"./peak_called/NA_model.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### finally we need some file transformation for rose to be able to process everything well\n",
    "\n",
    "we transform the outputed bed files into gff for rose to be able to process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import ROSE_utils\n",
    "ROSE_utils.bedToGFF(bed=\"/Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.bed\",output=\"/Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.gff\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we then sort and index the mapped reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 18 files and 6 in-memory blocks...\r\n"
     ]
    }
   ],
   "source": [
    "! samtools sort --threads 6 $mypath/mapped.sam > $mypath/mapped_sorted.bam "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we erase the now useless file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! rm $mypath/mapped.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we sort the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! samtools index -@ 7 $mypath/mapped_sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### finally we have to switch folders and then call rose\n",
    "rose will now stitch the enhancers to find the superenhancers in this cell line "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "folder /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/ does not exist\n",
      "folder /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/gff/ does not exist\n",
      "folder /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/ does not exist\n",
      "USING /Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.gff AS THE INPUT GFF\n",
      "USING hg19 AS THE GENOME\n",
      "MAKING START DICT\n",
      "LOADING IN GFF REGIONS\n",
      "CHECKING INPUT TO MAKE SURE EACH REGION HAS A UNIQUE IDENTIFIER\n",
      "REFERENCE COLLECTION PASSES QC\n",
      "STITCHING REGIONS TOGETHER\n",
      "PERFORMING REGION STITCHING\n",
      "MAKING GFF FROM STITCHED COLLECTION\n",
      "WRITING STITCHED GFF TO DISK AS /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/gff/NA_summits_12KB_STITCHED.gff\n",
      "OUTPUT WILL BE WRITTEN TO  /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/NA_summits_12KB_STITCHED_ENHANCER_REGION_MAP.txt\n",
      "python ROSE_bamToGFF.py -f 1 -e 200 -r -m 1 -b /Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam -i /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/gff/NA_summits_12KB_STITCHED.gff -o /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/NA_summits_12KB_STITCHED_mapped_sorted.bam_MAPPED.gff &\n",
      "python ROSE_bamToGFF.py -f 1 -e 200 -r -m 1 -b /Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam -i /Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.gff -o /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/NA_summits_mapped_sorted.bam_MAPPED.gff &\n",
      "PAUSING TO MAP\n",
      "{'matrix': '1', 'extension': '200', 'floor': '1', 'sense': 'both', 'output': '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/NA_summits_12KB_STITCHED_mapped_sorted.bam_MAPPED.gff', 'bam': '/Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam', 'rpm': True, 'input': '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/gff/NA_summits_12KB_STITCHED.gff'}\n",
      "[]\n",
      "mapping to GFF and making a matrix with fixed bin number\n",
      "{'matrix': '1', 'extension': '200', 'floor': '1', 'sense': 'both', 'output': '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/NA_summits_mapped_sorted.bam_MAPPED.gff', 'bam': '/Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam', 'rpm': True, 'input': '/Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.gff'}\n",
      "[]\n",
      "mapping to GFF and making a matrix with fixed bin number\n",
      "WAITING FOR MAPPING TO COMPLETE. ELAPSED TIME (MIN):\n",
      "0\n",
      "using a MMR value of 43.7552\n",
      "using a MMR value of 43.7552\n",
      "does not have chr\n",
      "does not have chr\n",
      "Number lines processed\n",
      "0\n",
      "Number lines processed\n",
      "0\n",
      "100\n",
      "100\n",
      "200\n",
      "200\n",
      "300\n",
      "300\n",
      "400\n",
      "400\n",
      "500\n",
      "500\n",
      "600\n",
      "600\n",
      "700\n",
      "700\n",
      "800\n",
      "800\n",
      "900\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1000\n",
      "1200\n",
      "1100\n",
      "1300\n",
      "1200\n",
      "1400\n",
      "1300\n",
      "1500\n",
      "1600\n",
      "1400\n",
      "1500\n",
      "1700\n",
      "1800\n",
      "1600\n",
      "1900\n",
      "1700\n",
      "2000\n",
      "1800\n",
      "2100\n",
      "2200\n",
      "1900\n",
      "2300\n",
      "2000\n",
      "2400\n",
      "2500\n",
      "2100\n",
      "2600\n",
      "2200\n",
      "2700\n",
      "2300\n",
      "2800\n",
      "2900\n",
      "2400\n",
      "3000\n",
      "2500\n",
      "3100\n",
      "2600\n",
      "3200\n",
      "2700\n",
      "3300\n",
      "2800\n",
      "3400\n",
      "2900\n",
      "3500\n",
      "3600\n",
      "3000\n",
      "3700\n",
      "3100\n",
      "3800\n",
      "3200\n",
      "3900\n",
      "4000\n",
      "3300\n",
      "4100\n",
      "3400\n",
      "4200\n",
      "3500\n",
      "4300\n",
      "4400\n",
      "3600\n",
      "4500\n",
      "3700\n",
      "4600\n",
      "3800\n",
      "4700\n",
      "3900\n",
      "4800\n",
      "4900\n",
      "4000\n",
      "5000\n",
      "4100\n",
      "5100\n",
      "5200\n",
      "4200\n",
      "5300\n",
      "4300\n",
      "5400\n",
      "4400\n",
      "5500\n",
      "5600\n",
      "4500\n",
      "5700\n",
      "4600\n",
      "5800\n",
      "4700\n",
      "5900\n",
      "6000\n",
      "4800\n",
      "6100\n",
      "4900\n",
      "6200\n",
      "5000\n",
      "6300\n",
      "6400\n",
      "5100\n",
      "6500\n",
      "5200\n",
      "6600\n",
      "5300\n",
      "6700\n",
      "6800\n",
      "5400\n",
      "6900\n",
      "5500\n",
      "7000\n",
      "5600\n",
      "7100\n",
      "7200\n",
      "5700\n",
      "7300\n",
      "5800\n",
      "7400\n",
      "5900\n",
      "7500\n",
      "7600\n",
      "6000\n",
      "7700\n",
      "6100\n",
      "7800\n",
      "7900\n",
      "6200\n",
      "8000\n",
      "6300\n",
      "8100\n",
      "6400\n",
      "8200\n",
      "6500\n",
      "8300\n",
      "8400\n",
      "6600\n",
      "8500\n",
      "6700\n",
      "8600\n",
      "6800\n",
      "8700\n",
      "8800\n",
      "6900\n",
      "8900\n",
      "7000\n",
      "9000\n",
      "9100\n",
      "7100\n",
      "9200\n",
      "7200\n",
      "9300\n",
      "7300\n",
      "9400\n",
      "9500\n",
      "7400\n",
      "9600\n",
      "7500\n",
      "9700\n",
      "7600\n",
      "9800\n",
      "9900\n",
      "7700\n",
      "10000\n",
      "7800\n",
      "10100\n",
      "7900\n",
      "10200\n",
      "10300\n",
      "8000\n",
      "10400\n",
      "8100\n",
      "10500\n",
      "8200\n",
      "10600\n",
      "8300\n",
      "10700\n",
      "10800\n",
      "8400\n",
      "10900\n",
      "8500\n",
      "11000\n",
      "8600\n",
      "11100\n",
      "11200\n",
      "8700\n",
      "11300\n",
      "8800\n",
      "11400\n",
      "8900\n",
      "11500\n",
      "11600\n",
      "9000\n",
      "11700\n",
      "9100\n",
      "11800\n",
      "9200\n",
      "11900\n",
      "12000\n",
      "9300\n",
      "12100\n",
      "9400\n",
      "12200\n",
      "9500\n",
      "12300\n",
      "12400\n",
      "9600\n",
      "12500\n",
      "9700\n",
      "12600\n",
      "9800\n",
      "12700\n",
      "12800\n",
      "9900\n",
      "12900\n",
      "10000\n",
      "13000\n",
      "10100\n",
      "13100\n",
      "10200\n",
      "13200\n",
      "13300\n",
      "10300\n",
      "13400\n",
      "10400\n",
      "13500\n",
      "10500\n",
      "13600\n",
      "10600\n",
      "13700\n",
      "13800\n",
      "10700\n",
      "13900\n",
      "10800\n",
      "14000\n",
      "10900\n",
      "14100\n",
      "11000\n",
      "14200\n",
      "14300\n",
      "11100\n",
      "14400\n",
      "11200\n",
      "14500\n",
      "11300\n",
      "14600\n",
      "14700\n",
      "11400\n",
      "14800\n",
      "11500\n",
      "14900\n",
      "11600\n",
      "15000\n",
      "15100\n",
      "11700\n",
      "15200\n",
      "11800\n",
      "15300\n",
      "11900\n",
      "15400\n",
      "12000\n",
      "15500\n",
      "15600\n",
      "12100\n",
      "15700\n",
      "15800\n",
      "12200\n",
      "15900\n",
      "12300\n",
      "16000\n",
      "16100\n",
      "12400\n",
      "16200\n",
      "12500\n",
      "16300\n",
      "16400\n",
      "12600\n",
      "16500\n",
      "12700\n",
      "16600\n",
      "16700\n",
      "12800\n",
      "16800\n",
      "12900\n",
      "16900\n",
      "13000\n",
      "17000\n",
      "17100\n",
      "13100\n",
      "17200\n",
      "13200\n",
      "17300\n",
      "13300\n",
      "17400\n",
      "17500\n",
      "13400\n",
      "17600\n",
      "13500\n",
      "17700\n",
      "13600\n",
      "17800\n",
      "17900\n",
      "13700\n",
      "18000\n",
      "13800\n",
      "18100\n",
      "13900\n",
      "18200\n",
      "18300\n",
      "14000\n",
      "18400\n",
      "14100\n",
      "18500\n",
      "14200\n",
      "18600\n",
      "18700\n",
      "14300\n",
      "18800\n",
      "14400\n",
      "18900\n",
      "19000\n",
      "14500\n",
      "19100\n",
      "14600\n",
      "19200\n",
      "14700\n",
      "19300\n",
      "19400\n",
      "14800\n",
      "19500\n",
      "14900\n",
      "19600\n",
      "15000\n",
      "19700\n",
      "19800\n",
      "15100\n",
      "19900\n",
      "15200\n",
      "20000\n",
      "15300\n",
      "20100\n",
      "20200\n",
      "15400\n",
      "20300\n",
      "15500\n",
      "20400\n",
      "15600\n",
      "20500\n",
      "20600\n",
      "15700\n",
      "20700\n",
      "15800\n",
      "20800\n",
      "15900\n",
      "20900\n",
      "21000\n",
      "16000\n",
      "21100\n",
      "16100\n",
      "21200\n",
      "16200\n",
      "21300\n",
      "21400\n",
      "16300\n",
      "21500\n",
      "16400\n",
      "21600\n",
      "16500\n",
      "21700\n",
      "21800\n",
      "16600\n",
      "21900\n",
      "16700\n",
      "22000\n",
      "16800\n",
      "22100\n",
      "22200\n",
      "16900\n",
      "22300\n",
      "17000\n",
      "22400\n",
      "17100\n",
      "22500\n",
      "22600\n",
      "17200\n",
      "22700\n",
      "17300\n",
      "22800\n",
      "17400\n",
      "22900\n",
      "23000\n",
      "17500\n",
      "23100\n",
      "17600\n",
      "23200\n",
      "17700\n",
      "23300\n",
      "23400\n",
      "17800\n",
      "23500\n",
      "17900\n",
      "23600\n",
      "23700\n",
      "23800\n",
      "23900\n",
      "24000\n",
      "24100\n",
      "24200\n",
      "24300\n",
      "24400\n",
      "24500\n",
      "24600\n",
      "24700\n",
      "24800\n",
      "24900\n",
      "25000\n",
      "25100\n",
      "25200\n",
      "25300\n",
      "25400\n",
      "25500\n",
      "25600\n",
      "25700\n",
      "25800\n",
      "25900\n",
      "26000\n",
      "26100\n",
      "26200\n",
      "26300\n",
      "26400\n",
      "26500\n",
      "26600\n",
      "26700\n",
      "26800\n",
      "MAPPING TOOK 25 MINUTES\n",
      "BAM MAPPING COMPLETED NOW MAPPING DATA TO REGIONS\n",
      "FORMATTING TABLE\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "GETTING MAPPED DATA\n",
      "GETTING MAPPING DATA FOR  /Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam\n",
      "OPENING /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/mappedGFF/NA_summits_12KB_STITCHED_mapped_sorted.bam_MAPPED.gff\n",
      "MAKING SIGNAL DICT FOR /Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam\n",
      "CALLING AND PLOTTING SUPER-ENHANCERS\n",
      "R --no-save /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/ /Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/NA_summits_12KB_STITCHED_ENHANCER_REGION_MAP.txt NA_summits NONE < ROSE_callSuper.R\n",
      "ARGUMENT '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/' __ignor'e__\n",
      "\n",
      "ARGUMENT '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/NA_summits_12KB_STITCHED_ENHANCER_REGION_MAP.txt' __ignor'e__\n",
      "\n",
      "ARGUMENT 'NA_summits' __ignor'e__\n",
      "\n",
      "ARGUMENT 'NONE' __ignor'e__\n",
      "\n",
      "\n",
      "R version 3.5.2 (2018-12-20) -- \"Eggshell Igloo\"\n",
      "Copyright (C) 2018 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-apple-darwin18.2.0 (64-bit)\n",
      "\n",
      "R est un logiciel libre livré sans AUCUNE GARANTIE.\n",
      "Vous pouvez le redistribuer sous certaines conditions.\n",
      "Tapez 'license()' ou 'licence()' pour plus de détails.\n",
      "\n",
      "R est un projet collaboratif avec de nombreux contributeurs.\n",
      "Tapez 'contributors()' pour plus d'information et\n",
      "'citation()' pour la façon de le citer dans les publications.\n",
      "\n",
      "Tapez 'demo()' pour des démonstrations, 'help()' pour l'aide\n",
      "en ligne ou 'help.start()' pour obtenir l'aide au format HTML.\n",
      "Tapez 'q()' pour quitter R.\n",
      "\n",
      "> #============================================================================\n",
      "> #==============SUPER-ENHANCER CALLING AND PLOTTING FUNCTIONS=================\n",
      "> #============================================================================\n",
      "> \n",
      ">  #This function calculates the cutoff by sliding a diagonal line and finding where it is tangential (or as close as possible)\n",
      ">  calculate_cutoff <- function(inputVector, drawPlot=TRUE,...){\n",
      "+  \tinputVector <- sort(inputVector)\n",
      "+ \tinputVector[inputVector<0]<-0 #set those regions with more control than ranking equal to zero\n",
      "+ \tslope <- (max(inputVector)-min(inputVector))/length(inputVector) #This is the slope of the line we want to slide. This is the diagonal.\n",
      "+ \txPt <- floor(optimize(numPts_below_line,lower=1,upper=length(inputVector),myVector= inputVector,slope=slope)$minimum) #Find the x-axis point where a line passing through that point has the minimum number of points below it. (ie. tangent)\n",
      "+ \ty_cutoff <- inputVector[xPt] #The y-value at this x point. This is our cutoff.\n",
      "+ \t\n",
      "+ \tif(drawPlot){  #if TRUE, draw the plot\n",
      "+ \t\tplot(1:length(inputVector), inputVector,type=\"l\",...)\n",
      "+ \t\tb <- y_cutoff-(slope* xPt)\n",
      "+ \t\tabline(v= xPt,h= y_cutoff,lty=2,col=8)\n",
      "+ \t\tpoints(xPt,y_cutoff,pch=16,cex=0.9,col=2)\n",
      "+ \t\tabline(coef=c(b,slope),col=2)\n",
      "+ \t\ttitle(paste(\"x=\",xPt,\"\\ny=\",signif(y_cutoff,3),\"\\nFold over Median=\",signif(y_cutoff/median(inputVector),3),\"x\\nFold over Mean=\",signif(y_cutoff/mean(inputVector),3),\"x\",sep=\"\"))\n",
      "+ \t\taxis(1,sum(inputVector==0),sum(inputVector==0),col.axis=\"pink\",col=\"pink\") #Number of regions with zero signal\n",
      "+ \t}\n",
      "+ \treturn(list(absolute=y_cutoff,overMedian=y_cutoff/median(inputVector),overMean=y_cutoff/mean(inputVector)))\n",
      "+ }\n",
      "> \n",
      "> #this is an accessory function, that determines the number of points below a diagnoal passing through [x,yPt]\n",
      "> numPts_below_line <- function(myVector,slope,x){\n",
      "+ \tyPt <- myVector[x]\n",
      "+ \tb <- yPt-(slope*x)\n",
      "+ \txPts <- 1:length(myVector)\n",
      "+ \treturn(sum(myVector<=(xPts*slope+b)))\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> convert_stitched_to_bed <- function(inputStitched,trackName,trackDescription,outputFile,splitSuper=TRUE,score=c(),superRows=c(),baseColor=\"0,0,0\",superColor=\"255,0,0\"){\n",
      "+ \toutMatrix <- matrix(data=\"\",ncol=4+ifelse(length(score)==nrow(inputStitched),1,0),nrow=nrow(inputStitched))\n",
      "+ \t\n",
      "+ \toutMatrix[,1] <- as.character(inputStitched$CHROM)\n",
      "+ \toutMatrix[,2] <- as.character(inputStitched$START)\n",
      "+ \toutMatrix[,3] <- as.character(inputStitched$STOP)\n",
      "+ \toutMatrix[,4] <- as.character(inputStitched$REGION_ID)\n",
      "+ \tif(length(score)==nrow(inputStitched)){\n",
      "+ \t\tscore <- rank(score,ties.method=\"first\")\n",
      "+ \t\tscore <- length(score)-score+1  #Stupid rank only does smallest to largest. \n",
      "+ \t\toutMatrix[,5] <- as.character(score)\n",
      "+ \t}\n",
      "+ \ttrackDescription <- paste(trackDescription,\"\\nCreated on \",format(Sys.time(), \"%b %d %Y\"),collapse=\"\",sep=\"\")\n",
      "+ \ttrackDescription <- gsub(\"\\n\",\"\\t\", trackDescription)\n",
      "+ \ttName <- gsub(\" \",\"_\",trackName)\n",
      "+ \tcat('track name=\"', tName,'\" description=\"', trackDescription,'\" itemRGB=On color=',baseColor,\"\\n\",sep=\"\",file=outputFile)\n",
      "+ \twrite.table(file= outputFile,outMatrix,sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE,append=TRUE)\n",
      "+ \tif(splitSuper==TRUE){\n",
      "+ \t\tcat(\"\\ntrack name=\\\"Super_\", tName,'\" description=\"Super ', trackDescription,'\" itemRGB=On color=', superColor,\"\\n\",sep=\"\",file=outputFile,append=TRUE)\n",
      "+ \t\twrite.table(file= outputFile,outMatrix[superRows,],sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE,append=TRUE)\n",
      "+ \t}\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> convert_stitched_to_gateway_bed <- function(inputStitched,outputFileRoot,splitSuper=TRUE,score=c(),superRows=c()){\n",
      "+ \toutMatrix <- matrix(data=\"\",ncol=6,nrow=nrow(inputStitched))\n",
      "+ \t\n",
      "+ \toutMatrix[,1] <- as.character(inputStitched$CHROM)\n",
      "+ \toutMatrix[,2] <- as.character(inputStitched$START)\n",
      "+ \toutMatrix[,3] <- as.character(inputStitched$STOP)\n",
      "+ \toutMatrix[,4] <- as.character(inputStitched$REGION_ID)\n",
      "+ \t\n",
      "+ \tif(length(score)==nrow(inputStitched)){\n",
      "+ \t\tscore <- rank(score,ties.method=\"first\")\n",
      "+ \t\tscore <- length(score)-score+1  #Stupid rank only does smallest to largest. \n",
      "+ \t\toutMatrix[,5] <- as.character(score)\n",
      "+ \t}\n",
      "+ \t\n",
      "+ \toutMatrix[,6] <- as.character(rep('.',nrow(outMatrix)))\n",
      "+ \t\n",
      "+ \t\n",
      "+ \toutputFile1 = paste(outputFileRoot,'_Gateway_Enhancers.bed',sep='')\n",
      "+ \twrite.table(file= outputFile1,outMatrix,sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE,append=TRUE)\n",
      "+ \tif(splitSuper==TRUE){\n",
      "+ \t\toutputFile2 = paste(outputFileRoot,'_Gateway_SuperEnhancers.bed',sep='')\n",
      "+ \n",
      "+ \t\twrite.table(file= outputFile2,outMatrix[superRows,],sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE,append=TRUE)\n",
      "+ \t}\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> writeSuperEnhancer_table <- function(superEnhancer,description,outputFile,additionalData=NA){\n",
      "+ \tdescription <- paste(\"#\",description,\"\\nCreated on \",format(Sys.time(), \"%b %d %Y\"),collapse=\"\",sep=\"\")\n",
      "+ \tdescription <- gsub(\"\\n\",\"\\n#\",description)\n",
      "+ \tcat(description,\"\\n\",file=outputFile)\n",
      "+ \tif(is.matrix(additionalData)){\n",
      "+ \t\tif(nrow(additionalData)!=nrow(superEnhancer)){\n",
      "+ \t\t\twarning(\"Additional data does not have the same number of rows as the number of super enhancers.\\n--->>> ADDITIONAL DATA NOT INCLUDED <<<---\\n\")\n",
      "+ \t\t}else{\n",
      "+ \t\t\tsuperEnhancer <- cbind(superEnhancer,additionalData)\n",
      "+ \t\t\tsuperEnhancer = superEnhancer[order(superEnhancer$enhancerRank),]\n",
      "+ \t\t\t\n",
      "+ \t\t}\n",
      "+ \t}\n",
      "+ \twrite.table(file=outputFile,superEnhancer,sep=\"\\t\",quote=FALSE,row.names=FALSE,append=TRUE)\n",
      "+ }\n",
      "> \n",
      "> \n",
      "> \n",
      "> #============================================================================\n",
      "> #===================SUPER-ENHANCER CALLING AND PLOTTING======================\n",
      "> #============================================================================\n",
      "> \n",
      "> \n",
      "> \n",
      "> args <- commandArgs()\n",
      "> \n",
      "> print('THESE ARE THE ARGUMENTS')\n",
      "[1] \"THESE ARE THE ARGUMENTS\"\n",
      "> print(args)\n",
      "[1] \"/usr/local/Cellar/r/3.5.2_1/lib/R/bin/exec/R\"                                                                    \n",
      "[2] \"--no-save\"                                                                                                       \n",
      "[3] \"/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/\"                                                \n",
      "[4] \"/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/NA_summits_12KB_STITCHED_ENHANCER_REGION_MAP.txt\"\n",
      "[5] \"NA_summits\"                                                                                                      \n",
      "[6] \"NONE\"                                                                                                            \n",
      "> \n",
      "> #ARGS\n",
      "> outFolder = args[3]\n",
      "> enhancerFile = args[4]\n",
      "> enhancerName = args[5]\n",
      "> wceName = args[6]\n",
      "> \n",
      "> \n",
      "> \n",
      "> #Read enhancer regions with closestGene columns\n",
      "> stitched_regions <- read.delim(file= enhancerFile,sep=\"\\t\")\n",
      "> \n",
      "> \n",
      "> #perform WCE subtraction. Using pipeline table to match samples to proper background. \n",
      "> rankBy_factor = colnames(stitched_regions)[7]\n",
      "> \n",
      "> prefix = unlist(strsplit(rankBy_factor,'_'))[1]\n",
      "> \n",
      "> if(wceName == 'NONE'){\n",
      "+ \t\n",
      "+ \t\n",
      "+ \trankBy_vector = as.numeric(stitched_regions[,7])\n",
      "+ \t\n",
      "+ }else{\n",
      "+ \twceName = colnames(stitched_regions)[8]\n",
      "+ \tprint('HERE IS THE WCE NAME')\n",
      "+ \tprint(wceName)\n",
      "+ \t\n",
      "+ \trankBy_vector = as.numeric(stitched_regions[,7])-as.numeric(stitched_regions[,8])\n",
      "+ }\t\n",
      "> \t\n",
      "> \n",
      "> #SETTING NEGATIVE VALUES IN THE rankBy_vector to 0\n",
      "> \n",
      "> rankBy_vector[rankBy_vector < 0] <- 0\n",
      "> \n",
      "> \n",
      "> #FIGURING OUT THE CUTOFF\n",
      "> \n",
      "> cutoff_options <- calculate_cutoff(rankBy_vector, drawPlot=FALSE,xlab=paste(rankBy_factor,'_enhancers'),ylab=paste(rankByFactor,' Signal','- ',wceName),lwd=2,col=4)\n",
      "> \n",
      "> \n",
      "> #These are the super-enhancers\n",
      "> superEnhancerRows <- which(rankBy_vector> cutoff_options$absolute)\n",
      "> typicalEnhancers = setdiff(1:nrow(stitched_regions),superEnhancerRows)\n",
      "> enhancerDescription <- paste(enhancerName,\" Enhancers\\nCreated from \", enhancerFile,\"\\nRanked by \",rankBy_factor,\"\\nUsing cutoff of \",cutoff_options$absolute,\" for Super-Enhancers\",sep=\"\",collapse=\"\")\n",
      "> \n",
      "> \n",
      "> #MAKING HOCKEY STICK PLOT\n",
      "> plotFileName = paste(outFolder,enhancerName,'_Plot_points.png',sep='')\n",
      "> png(filename=plotFileName,height=600,width=600)\n",
      "> signalOrder = order(rankBy_vector,decreasing=TRUE)\n",
      "> if(wceName == 'NONE'){\n",
      "+ \tplot(length(rankBy_vector):1,rankBy_vector[signalOrder], col='red',xlab=paste(rankBy_factor,'_enhancers'),ylab=paste(rankBy_factor,' Signal'),pch=19,cex=2)\t\n",
      "+ \t\n",
      "+ }else{\n",
      "+ \tplot(length(rankBy_vector):1,rankBy_vector[signalOrder], col='red',xlab=paste(rankBy_factor,'_enhancers'),ylab=paste(rankBy_factor,' Signal','- ',wceName),pch=19,cex=2)\n",
      "+ }\n",
      "> abline(h=cutoff_options$absolute,col='grey',lty=2)\n",
      "> abline(v=length(rankBy_vector)-length(superEnhancerRows),col='grey',lty=2)\n",
      "> lines(length(rankBy_vector):1,rankBy_vector[signalOrder],lwd=4, col='red')\n",
      "> text(0,0.8*max(rankBy_vector),paste(' Cutoff used: ',cutoff_options$absolute,'\\n','Super-Enhancers identified: ',length(superEnhancerRows)),pos=4)\n",
      "> \n",
      "> dev.off()\n",
      "null device \n",
      "          1 \n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> #Writing a bed file\n",
      "> bedFileName = paste(outFolder,enhancerName,'_Enhancers_withSuper.bed',sep='')\n",
      "> convert_stitched_to_bed(stitched_regions,paste(rankBy_factor,\"Enhancers\"), enhancerDescription,bedFileName,score=rankBy_vector,splitSuper=TRUE,superRows= superEnhancerRows,baseColor=\"0,0,0\",superColor=\"255,0,0\")\n",
      "> \n",
      "> \n",
      "> bedFileRoot = paste(outFolder,enhancerName,sep='')\n",
      "> \n",
      "> convert_stitched_to_gateway_bed(stitched_regions,bedFileRoot,splitSuper=TRUE,score=rankBy_vector,superRows = superEnhancerRows)\n",
      "> \n",
      "> \n",
      "> #This matrix is just the super_enhancers\n",
      "> true_super_enhancers <- stitched_regions[superEnhancerRows,]\n",
      "> \n",
      "> additionalTableData <- matrix(data=NA,ncol=2,nrow=nrow(stitched_regions))\n",
      "> colnames(additionalTableData) <- c(\"enhancerRank\",\"isSuper\")\n",
      "> additionalTableData[,1] <- nrow(stitched_regions)-rank(rankBy_vector,ties.method=\"first\")+1\n",
      "> additionalTableData[,2] <- 0\n",
      "> additionalTableData[superEnhancerRows,2] <- 1\n",
      "> \n",
      "> \n",
      "> #Writing enhancer and super-enhancer tables with enhancers ranked and super status annotated\n",
      "> enhancerTableFile = paste(outFolder,enhancerName,'_AllEnhancers.table.txt',sep='')\n",
      "> writeSuperEnhancer_table(stitched_regions, enhancerDescription,enhancerTableFile, additionalData= additionalTableData)\n",
      "Warning message:\n",
      "In write.table(file = outputFile, superEnhancer, sep = \"\\t\", quote = FALSE,  :\n",
      "  appending column names to file\n",
      "> \n",
      "> superTableFile = paste(outFolder,enhancerName,'_SuperEnhancers.table.txt',sep='')\n",
      "> writeSuperEnhancer_table(true_super_enhancers, enhancerDescription,superTableFile, additionalData= additionalTableData[superEnhancerRows,])\n",
      "Warning message:\n",
      "In write.table(file = outputFile, superEnhancer, sep = \"\\t\", quote = FALSE,  :\n",
      "  appending column names to file\n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> \n",
      "> \t\n",
      "> \n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "a = ['/Users/jeremie/Documents/Projects/BroadInstitute'] #! pwd\n",
    "a = a[0]\n",
    "path_to_rose = \"/Users/jeremie/Documents/logiciels/rose\"\n",
    "os.chdir(path_to_rose)\n",
    "! python ROSE_main.py -g hg19 -i $a/peak_called/NA_summits.gff -r $a/$mypath/mapped_sorted.bam -o $a/superenhancers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### We can now create a map of the core regulatory circuitry of this cell line!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "they were two lines to change in crcmapper to make it work...\n",
    "```python\n",
    "def createExpressionDict(annotationFile, projectFolder, projectName, refseqToNameDict, expressionTable):\n",
    "```\n",
    "line 98\n",
    "\n",
    "and \n",
    "```python\n",
    "for line in expressionTable[1:]:\n",
    "    trid = line[0]\n",
    "    geneName = refseqToNameDict[trid]\n",
    "    if len(expressionTable[1]) == 3:  # when expressionTable is an output from bamToGFF.py\n",
    "        exp = float(line[2])\n",
    "    else:  # when expressionTable is passed as an option (2 columns)\n",
    "        exp = float(line[1])\n",
    "```\n",
    "line 120 \n",
    "\n",
    "creating syntax error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'enhancers': '/Users/jeremie/Documents/Projects/BroadInstitute/superenhancers/NA_summits_AllEnhancers.table.txt', 'name': 'CRC', 'extension': 500, 'expCutoff': 33, 'subpeaks': '/Users/jeremie/Documents/Projects/BroadInstitute/peak_called/NA_summits.bed', 'Enumber': 'supers', 'genome': 'HG19', 'activity': None, 'output': '/Users/jeremie/Documents/Projects/BroadInstitute/crcmapper', 'fasta': '/Users/jeremie/Documents/Projects/BroadInstitute/datatosend/reference.fa', 'bam': '/Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam'}\n",
      "IDENTIFY EXPRESSED GENES\n",
      "{'unique': 0, 'matrix': '1', 'extension': '200', 'floor': '0', 'input': '/Users/jeremie/Documents/Projects/BroadInstitute/crcmapperCRC_TSS.gff', 'density': True, 'cluster': None, 'bam': '/Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam', 'sense': 'both', 'output': '/Users/jeremie/Documents/Projects/BroadInstitute/crcmappermatrix.gff', 'total': False, 'rpm': True, 'jxn': False}\n",
      "[]\n",
      "mapping to GFF and making a matrix with fixed bin number\n",
      "using a MMR value of 43.7552\n",
      "Number lines processed\n",
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "3600\n",
      "3700\n",
      "3800\n",
      "3900\n",
      "4000\n",
      "4100\n",
      "4200\n",
      "4300\n",
      "4400\n",
      "4500\n",
      "4600\n",
      "4700\n",
      "4800\n",
      "4900\n",
      "5000\n",
      "5100\n",
      "5200\n",
      "5300\n",
      "5400\n",
      "5500\n",
      "5600\n",
      "5700\n",
      "5800\n",
      "5900\n",
      "6000\n",
      "6100\n",
      "6200\n",
      "6300\n",
      "6400\n",
      "6500\n",
      "6600\n",
      "6700\n",
      "6800\n",
      "6900\n",
      "7000\n",
      "7100\n",
      "7200\n",
      "7300\n",
      "7400\n",
      "7500\n",
      "7600\n",
      "7700\n",
      "7800\n",
      "7900\n",
      "8000\n",
      "8100\n",
      "8200\n",
      "8300\n",
      "8400\n",
      "8500\n",
      "8600\n",
      "8700\n",
      "8800\n",
      "8900\n",
      "9000\n",
      "9100\n",
      "9200\n",
      "9300\n",
      "9400\n",
      "9500\n",
      "9600\n",
      "9700\n",
      "9800\n",
      "9900\n",
      "10000\n",
      "10100\n",
      "10200\n",
      "10300\n",
      "10400\n",
      "10500\n",
      "10600\n",
      "10700\n",
      "10800\n",
      "10900\n",
      "11000\n",
      "11100\n",
      "11200\n",
      "11300\n",
      "11400\n",
      "11500\n",
      "11600\n",
      "11700\n",
      "11800\n",
      "11900\n",
      "12000\n",
      "12100\n",
      "12200\n",
      "12300\n",
      "12400\n",
      "12500\n",
      "12600\n",
      "12700\n",
      "12800\n",
      "12900\n",
      "13000\n",
      "13100\n",
      "13200\n",
      "13300\n",
      "13400\n",
      "13500\n",
      "13600\n",
      "13700\n",
      "13800\n",
      "13900\n",
      "14000\n",
      "14100\n",
      "14200\n",
      "14300\n",
      "14400\n",
      "14500\n",
      "14600\n",
      "14700\n",
      "14800\n",
      "14900\n",
      "15000\n",
      "15100\n",
      "15200\n",
      "15300\n",
      "15400\n",
      "15500\n",
      "15600\n",
      "15700\n",
      "15800\n",
      "15900\n",
      "16000\n",
      "16100\n",
      "16200\n",
      "16300\n",
      "16400\n",
      "16500\n",
      "16600\n",
      "16700\n",
      "16800\n",
      "16900\n",
      "17000\n",
      "17100\n",
      "17200\n",
      "17300\n",
      "17400\n",
      "17500\n",
      "17600\n",
      "17700\n",
      "17800\n",
      "17900\n",
      "18000\n",
      "18100\n",
      "18200\n",
      "18300\n",
      "18400\n",
      "18500\n",
      "18600\n",
      "18700\n",
      "18800\n",
      "18900\n",
      "19000\n",
      "19100\n",
      "19200\n",
      "19300\n",
      "19400\n",
      "19500\n",
      "19600\n",
      "19700\n",
      "19800\n",
      "19900\n",
      "20000\n",
      "20100\n",
      "20200\n",
      "20300\n",
      "20400\n",
      "20500\n",
      "20600\n",
      "20700\n",
      "20800\n",
      "20900\n",
      "21000\n",
      "21100\n",
      "21200\n",
      "21300\n",
      "21400\n",
      "21500\n",
      "21600\n",
      "21700\n",
      "21800\n",
      "21900\n",
      "22000\n",
      "22100\n",
      "22200\n",
      "22300\n",
      "22400\n",
      "22500\n",
      "22600\n",
      "22700\n",
      "22800\n",
      "22900\n",
      "23000\n",
      "23100\n",
      "23200\n",
      "23300\n",
      "23400\n",
      "23500\n",
      "23600\n",
      "23700\n",
      "23800\n",
      "23900\n",
      "24000\n",
      "24100\n",
      "24200\n",
      "24300\n",
      "24400\n",
      "24500\n",
      "24600\n",
      "24700\n",
      "24800\n",
      "24900\n",
      "25000\n",
      "25100\n",
      "25200\n",
      "25300\n",
      "25400\n",
      "25500\n",
      "25600\n",
      "25700\n",
      "25800\n",
      "python ./bamToGFF.py -r  -d  -o /Users/jeremie/Documents/Projects/BroadInstitute/crcmappermatrix.gff -m 1 -f 0 -e 200  -i /Users/jeremie/Documents/Projects/BroadInstitute/crcmapperCRC_TSS.gff -b /Users/jeremie/Documents/Projects/BroadInstitute/datatosend/mapped_sorted.bam\n",
      "CREATING SUPER-ENHANCER LOCUS COLLECTION\n",
      "CREATING EXPRESSION DICTIONARY\n",
      "Expression cutoff: 0.0\n",
      "FINDING CANIDATE TFs\n",
      "Number of canidate TFs: 0\n",
      "CREATE CANDIDATE TFs AND SE TABLE\n",
      "MAKE FASTA\n",
      "MOTIF SEARCH\n",
      "Number of annotated candidate TFs that have motifs: 0\n",
      "/bin/sh: fasta-get-markov: command not found\n",
      "fimo -verbosity 1 -text -oc /Users/jeremie/Documents/Projects/BroadInstitute/crcmapper --bgfile /Users/jeremie/Documents/Projects/BroadInstitute/crcmapperCRC_bg.meme ./VertebratePWMs.txt /Users/jeremie/Documents/Projects/BroadInstitute/crcmapperCRC_SUBPEAKS.fa > /Users/jeremie/Documents/Projects/BroadInstitute/crcmapperfimo.txt\n",
      "/bin/sh: fimo: command not found\n",
      "IDENTIFY TF-TF INTERACTIONS\n",
      "Traceback (most recent call last):\n",
      "  File \"CRCmapper.py\", line 613, in <module>\n",
      "    main()\n",
      "  File \"CRCmapper.py\", line 601, in main\n",
      "    graph = buildNetwork(projectFolder, projectName, candidateGenes, refseqToNameDict, motifConvertFile)\n",
      "  File \"CRCmapper.py\", line 373, in buildNetwork\n",
      "    fimoTable = utils.parseTable(fimoFile, '\\t')\n",
      "  File \"/Users/jeremie/Documents/logiciels/crcmapper/utils.py\", line 245, in parseTable\n",
      "    if lines[0].count('\\r') > 0:\n",
      "IndexError: list index out of range\n"
     ]
    }
   ],
   "source": [
    "path_to_crcmapper = \"/Users/jeremie/Documents/logiciels/crcmapper\"\n",
    "os.chdir(path_to_crcmapper)\n",
    "\n",
    "! python CRCmapper.py -e $a/superenhancers/NA_summits_AllEnhancers.table.txt -b $a/$mypath/mapped_sorted.bam -g HG19 -f $a/$mypath/reference -s $a/peak_called/NA_summits.bed -n CRC -o $a/crcmapper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## an automated version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mypath = \"datatosend\"\n",
    "refs = [30,31,32,33,34]\n",
    "files = {}\n",
    "path_to_rose = \"/Users/jeremie/Documents/logiciels/rose\"\n",
    "path_to_crcmapper = \"/Users/jeremie/Documents/logiciels/crcmapper\"\n",
    "a = ! pwd\n",
    "a = a[0]\n",
    "os.chdir(path_to_rose)\n",
    "import ROSE_utils\n",
    "os.chdir(a)\n",
    "for ref in refs\n",
    "    for r in os.listdir(mypath):\n",
    "        if r[:2] in refs:\n",
    "            lane = \"R1\" if \"_R1\" in r else \"R2\"\n",
    "            files[lane] = r\n",
    "            ! bowtie2 -x index -1 $mypath/$files[\"R1\"] -2 $mypath/$files[\"R2\"] -S $mypath/mapped.sam --threads 8 -t\n",
    "            ! MACS2 callpeak -t $mypath/mapped.sam --outdir peak_called_$ref\n",
    "            ! Rscript peak_called_$ref/NA_model.r\n",
    "            ! samtools sort --threads 6 $mypath/mapped.sam > $mypath/mapped_sorted_$ref.bam \n",
    "            ! rm $mypath/mapped.sam\n",
    "            ! samtools index -@ 7 $mypath/mapped_sorted_$ref.bam \n",
    "            os.chdir(path_to_rose)\n",
    "            ROSE_utils.bedToGFF(bed=a + \"/peak_called_\"+ref+\"/NA_summits.bed\",output=a + \"/peak_called_\"+ ref +\"/NA_summits.gff\")\n",
    "            ! python ROSE_main.py -g hg19 -i $a/peak_called_$ref/NA_summits.gff -r $a/$mypath/mapped_sorted_$ref.bam -o $a/superenhancers_$ref\n",
    "            os.chdir(path_to_crcmapper)\n",
    "            ! python CRCmapper.py -e $a/superenhancers_$ref/NA_summits_AllEnhancers.table.txt -b $a/$mypath/mapped_sorted_$ref.bam -g hg19 -f $a/$mypath/reference  -s $a/peak_called_$ref/NA_summits.bed -n CRC -o $a/crcmapper_$ref\n",
    "            os.chdir(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bed_to_gff(input_name, output_name): # thanks to glaaxy\n",
    "    skipped_lines = 0\n",
    "    first_skipped_line = 0\n",
    "    out = open( output_name, 'w' )\n",
    "    out.write( \"##gff-version 2\\n\" )\n",
    "    out.write( \"##bed_to_gff_converter.py\\n\\n\" )\n",
    "    i = 0\n",
    "    for i, line in enumerate( file( input_name ) ):\n",
    "        complete_bed = False\n",
    "        line = line.rstrip( '\\r\\n' )\n",
    "        if line and not line.startswith( '#' ) and not line.startswith( 'track' ) and not line.startswith( 'browser' ):\n",
    "            try:\n",
    "                elems = line.split( '\\t' )\n",
    "                if len( elems ) == 12:\n",
    "                    complete_bed = True\n",
    "                chrom = elems[0]\n",
    "                if complete_bed:\n",
    "                    feature = \"mRNA\"\n",
    "                else:\n",
    "                    try:\n",
    "                        feature = elems[3]\n",
    "                    except:\n",
    "                        feature = 'feature%d' % ( i + 1 )\n",
    "                start = int( elems[1] ) + 1\n",
    "                end = int( elems[2] )\n",
    "                try:\n",
    "                    score = elems[4]\n",
    "                except:\n",
    "                    score = '0'\n",
    "                try:\n",
    "                    strand = elems[5]\n",
    "                except:\n",
    "                    strand = '+'\n",
    "                try:\n",
    "                    group = elems[3]\n",
    "                except:\n",
    "                    group = 'group%d' % ( i + 1 )\n",
    "                if complete_bed:\n",
    "                    out.write( 'chr%s\\t%s\\t\\t%d\\t%d\\t\\t\\t%s\\t.\\t\\t%s %s;\\n' % ( chrom, feature, start, end, score, strand, feature, group  ) )\n",
    "                else:\n",
    "                    out.write( 'chr%s\\t%s\\t\\t%d\\t%d\\t%s\\t%s\\t.\\t\\t%s;\\n' % ( chrom, feature, start, end, score, strand, group  ) )\n",
    "                if complete_bed:\n",
    "                    # We have all the info necessary to annotate exons for genes and mRNAs\n",
    "                    block_count = int( elems[9] )\n",
    "                    block_sizes = elems[10].split( ',' )\n",
    "                    block_starts = elems[11].split( ',' )\n",
    "                    for j in range( block_count ):\n",
    "                        exon_start = int( start ) + int( block_starts[j] )\n",
    "                        exon_end = exon_start + int( block_sizes[j] ) - 1\n",
    "                        out.write( '%s\\tbed2gff\\texon\\t%d\\t%d\\t%s\\t%s\\t.\\texon %s;\\n' % ( chrom, exon_start, exon_end, score, strand, group ) )\n",
    "            except:\n",
    "                skipped_lines += 1\n",
    "                if not first_skipped_line:\n",
    "                    first_skipped_line = i + 1\n",
    "        else:\n",
    "            skipped_lines += 1\n",
    "            if not first_skipped_line:\n",
    "                first_skipped_line = i + 1\n",
    "    out.close()\n",
    "    info_msg = \"%i lines converted to GFF version 2.  \" % ( i + 1 - skipped_lines )\n",
    "    if skipped_lines > 0:\n",
    "        info_msg += \"Skipped %d blank/comment/invalid lines starting with line #%d.\" %( skipped_lines, first_skipped_line )\n",
    "    print info_msg"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
